In [12]:
"""
These tests are an important sanity check

If any modification ruins the test functionality then we can be sure 
that the modification is detrimental
"""

import numpy as np
import matplotlib.pyplot as plt

# Add Backend to path
import sys

"""
Change Path according to own machine
"""

#sys.path.append('C:/Users/eeveetza/Documents/GitHub/AlgoBrain/src/com/company/Backend')
sys.path.append('C:/Users/alexa/OneDrive/Documents/GitHub/AlgoBrain/src/com/company/Backend')
#sys.path.append('C:/Users/Tbuob/OneDrive/Dokumente/GitHub/AlgoBrain/src/com/company/Backend')

from Environment import World
from Experiment import Experiment

"""
Mapping integers - instructions:
    
0  <==> NopA
1  <==> NopB
2  <==> NopC
3  <==> IfNEq
4  <==> IfLess
5  <==> Swap
6  <==> Pop
7  <==> Push
8  <==> SwapStack
9  <==> RightShift
10 <==> LeftShift
11 <==> Inc
12 <==> Dec
13 <==> Add
14 <==> Sub
15 <==> Nand
16 <==> HAlloc
17 <==> HDivide
18 <==> IO
19 <==> HCopy
20 <==> HSearch
21 <==> MovHead
22 <==> JmpHead
23 <==> GetHead£
24 <==> SetFlow
25 <==> IfLabel
26 <==> Move
27 <==> Sexual Reproduction

"""

# clears all registers (under the assumption that the active stack is empty)
clear = [6, 0, 6, 1, 6, 2]

# Custom made test instructions
program_not = [12, 1, 18, 2, 15, 0, 18, 0] + clear
program_nand = [18,1,18,2,15,0,18,0] + clear
program_and = [18,1,18,2,15,2,6,1,14,0,12,0,18,0] + clear
program_or = [6, 1, 18, 2, 14, 0, 12, 0, 6, 1, 18, 2, 14, 2, 12, 2, 7, 0, 6, 1, 15, 0, 18, 0] + clear
program_or_n = [6, 1, 18, 2, 14, 2, 12, 2, 18, 1, 15, 0, 18, 0] + clear
program_and_n = [6, 1, 18, 2, 14, 2, 12, 2, 18, 1, 15, 2, 6, 1, 14, 0, 12, 0, 18, 0] + clear
program_nor = [18, 2, 14, 0, 12, 0, 6, 1, 18, 2, 14, 2, 12, 2, 7, 0, 6, 1, 15, 2, 6, 1, 14, 0, 12, 0, 18, 0] + clear
program_xor = [18, 0, 18, 2, 7, 0, 7, 2, 8, 14, 2, 12, 2, 7, 0, 6, 1, 6, 0, 15,\
           2, 6, 1, 7, 2, 8, 6, 0, 6, 2, 14, 2, 12, 2, 7, 0, 6, 1, 6, 0, 15, 2, 8, 6, 1, 15, 0, 18, 0] + clear
program_equ = [18, 1, 18, 2, 7, 1, 7, 2, 8, 15, 2, 6, 1, 7, 2, 8, 6, 0, 6, 2, 14, 2, 12, 2, 7,\
               2, 7, 0, 6, 2, 14, 2, 12, 2, 6, 1, 15, 2, 8, 6, 1, 15, 0, 18, 0]

In [11]:
len(program_equ)

45

In [34]:
#%%
"""
Parser to create instructions by reading input stream

All instructions to be written lowercase without dashes
"""

result = []
instr = []
while True:
    data = input("Instructions: ")
    if data == "end":
            print("result = " + str(result))
            break
    else:
        number_of_empty_responses = 0
        
        # Boring case checking
        if data == "nopa" or data == "a":
            result.append(0)
            instr.append("nopa")
        elif data == "nopb" or data == "b":
            result.append(1)
            instr.append("nopb")
        elif data == "nopc" or data == "c":
            result.append(2)
            instr.append("nopc")
        elif data == "ifneq":
            result.append(3)
        elif data == "ifless":
            result.append(4)
        elif data == "swap":
            result.append(5)
        elif data == "pop":
            result.append(6)
        elif data == "push":
            result.append(7)
        elif data == "swapstack":
            result.append(8)
        elif data == "rshift":
            result.append(9)
        elif data == "lshift":
            result.append(10)
        elif data == "inc":
            result.append(11)
        elif data == "dec":
            result.append(12)
        elif data == "add":
            result.append(13)
        elif data == "sub":
            result.append(14)
        elif data == "nand":
            result.append(15)
        elif data == "halloc":
            result.append(16)
        elif data == "hdivide":
            result.append(17)
        elif data == "io":
            result.append(18)
        elif data == "hcopy":
            result.append(19)
        elif data == "hsearch":
            result.append(20)
        elif data == "movhead":
            result.append(21)
        elif data == "jmphead":
            result.append(22)
        elif data == "gethead":
            result.append(23)
        elif data == "setflow":
            result.append(24)
        elif data == "iflabel":
            result.append(25)
        elif data == "consume":
            result.append(26)
        elif data == "back":
            result.pop(-1)
        elif data == "clear":
            result = []
        print(result)

Instructions: dec
[12]
Instructions: b
[12, 1]
Instructions: io
[12, 1, 18]
Instructions: c
[12, 1, 18, 2]
Instructions: nand
[12, 1, 18, 2, 15]
Instructions: a
[12, 1, 18, 2, 15, 0]
Instructions: io
[12, 1, 18, 2, 15, 0, 18]
Instructions: a
[12, 1, 18, 2, 15, 0, 18, 0]
Instructions: end
result = [12, 1, 18, 2, 15, 0, 18, 0]


In [3]:
#%%
"""
Test 1:
    
    An organism that computes NOT
    
    NOT(A) == -A - 1
    
    NOT(A) = NAND(A, 1111...111)

    DEC-B
    IO-C
    NAND-A
    IO-A
    
"""
instructions = program_not

world = World(1, replacement_strategy = "neighborhood", cm_prob = 0, ins_prob = 0, del_prob = 0,notify_ = True)

world.place_custom(program_not)

world.schedule(50)


EMULATOR AT POSITION (0, 0) COMPUTED NOT
INPUT:  1335684869
OUTPUT: 2959282426

EMULATOR AT POSITION (0, 0) COMPUTED NOT
INPUT:  1873646631
OUTPUT: 2421320664


In [38]:
#%%
"""
Test 2:
    
    An organim that computes NAND
    
    IO-B
    IO-C
    NAND-A
    IO-A
    
"""
world = World(1, replacement_strategy = "neighborhood", cm_prob = 0, ins_prob = 0, del_prob = 0, notify_ = True)

world.place_custom(program_nand)

world.schedule(60)


EMULATOR AT POSITION (0, 0) COMPUTED NAND
INPUT: [ 322259418 1206126325]
OUTPUT:  4242276143


In [39]:
"""
Test 3:
    
    An organism that computes AND
    
    Serial connection NAND => NOT
    
    AND(A,B) = NOT(NAND(A,B))
    
    IO-B
    IO-C
    NAND-C
    POP-B
    SUB-A
    INC-A
    IO-A
    
"""
world = World(1, replacement_strategy = "neighborhood", cm_prob = 0, ins_prob = 0, del_prob = 0, notify_ = True)

world.place_custom(program_and)

world.schedule(500)


EMULATOR AT POSITION (0, 0) COMPUTED AND
INPUT: [4000955396 1529749002]
OUTPUT:  1244142592


In [40]:
world.pool.get((0,0)).fun_and

True

In [42]:
"""
Test 4:
    
    An organism that computes OR
    
    OR(A,B) == NOT(NOT(A) AND NOT(B))
    
    NOT(A) AND NOT(B) = NOT(NAND(NOT(A),NOT(B)))
    
    All together:
    
        OR(A,B) = NAND(NOT(A),NOT(B))
        
    
    #1: 
        Save NOT(A) in regA
        
        POP-B
        IO-C
        SUB-A
        DEC-A
        
    #2:
        Save NOT(B) in regC
        
        POP-B
        IO-C
        SUB-C
        DEC-C
        
    #3:
        Move regA to regB
        
        PUSH-A
        POP-B
    
    #4:
        NAND to regA, IO-A
        
        NAND-A
        IO-A
"""
world = World(1, replacement_strategy = "neighborhood", cm_prob = 0, ins_prob = 0, del_prob = 0, notify_ = True)

world.place_custom(program_or)

world.schedule(30)


EMULATOR AT POSITION (0, 0) COMPUTED OR
INPUT: [2884037911 3420455308]
OUTPUT:  3957779871


In [43]:
"""
Test 5:
    
    An organism that computes OR_N
    
    Take A OR NOT(B)
    
    A OR NOT(B) = NOT(A) NAND B
    
    # Save NOT(A) to C
    # Input B
    # NAND to A
    # IO-A
    
    
"""
world = World(1, replacement_strategy = "neighborhood", cm_prob = 0, ins_prob = 0, del_prob = 0, notify_ = True)

world.place_custom(program_or_n)

world.schedule(30)


EMULATOR AT POSITION (0, 0) COMPUTED OR_N
INPUT: [2010218037 3123351556]
OUTPUT:  3123614158


In [45]:
"""
Test 6:
    
    An organism that computes AND_N
    
    Take A AND NOT(B)
    
    a and not(b) = not(a nand not(b))
    
    # Save NOT(B) to regC
    # Load A to regB
    # NAND-C
    # POP-B
    # SUB-A
    # DEC-A
    # IO-A
    
"""
world = World(1, replacement_strategy = "neighborhood", cm_prob = 0, ins_prob = 0, del_prob = 0, notify_ = True)

world.place_custom(program_and_n)

world.schedule(30)


EMULATOR AT POSITION (0, 0) COMPUTED AND_N
INPUT: [ 905126097 3262851483]
OUTPUT:  897581120


In [55]:
"""
Test 7:
    
    An organism that computes NOR
    
    NOT(A OR B)
    
    A OR B = NAND(NOT(A),NOT(B))
    
    How to do it:
        
    #1: 
        Save NOT(A) in regA
        
        IO-C
        SUB-A
        DEC-A
        
    #2:
        Save NOT(B) in regC
        
        POP-B
        IO-C
        SUB-C
        DEC-C
        
    #3:
        Move regA to regB
        
        PUSH-A
        POP-B
    
    #4:
        NAND to regC
        
        NAND-C
        
    #5: Clear B
        POP-B
        
    #6: Sub to A, Dec A, IO A
    
"""
world = World(1, replacement_strategy = "neighborhood", cm_prob = 0, ins_prob = 0, del_prob = 0, notify_ = True)

world.place_custom(program_nor)

world.schedule(40)


EMULATOR AT POSITION (0, 0) COMPUTED NOR
INPUT: [  90499245 3810500182]
OUTPUT:  411042048


In [56]:
"""
Test 8:
    
    An organism that computes XOR
    
    (A and ~B) or (~A and B)
    
    NOT(NAND(A,NOT(B))) OR NOT(NAND(NOT(A),B))
    
    This ends up being equivalent to
    
    nand(nand(a,not(b)),nand(not(a),b))
    
    #1:
        input one to a
        input other to c
        push both to stack for later use. push 2nd input first and then the first one
        switch stack for popping
        calculate not(other) as:
            sub-c
            dec-c
        move first one from a to b by doing:
            push-a
            pop-b
            pop-a
        
        now we have:
            reg_a = 0
            reg_b = Input0
            reg_c = not(Input1)
            stack = (b,a)
        we just nand this to c
        pop-b
        push-c
        
        repeat first part
        
        swap stack
        pop a
        pop c
        
        Detailed Guide:
            
            # Inputing and saving to stack:
                IO-A
                IO-C
                PUSH-A
                PUSH-C
                
            Stack0 now is [i1,i0]
            
            Registers are:
                A = i0
                B = 0
                C = i1
                
            swapstack
            
            # Calculating not(i1):
                sub-c
                dec-c
                
            # Move one from A to B:
                push-a
                pop-b
                pop-a
                
                
            Registers are:
                
                regA = 0
                regB = A
                regC = NOT(B)
                
            # Nand them to C:
                
                nand-c
                
            # pop b to clear it:
                pop-b
            
            push-c
            
            now the second stack has nand(a,not(b)) in it.
            Switching to first stack with [b,a] in it
            
            swapstack
            pop-a
            pop-c
            
            # do not of register C which contains A
            
            sub-c
            dec-c
            
            # move B from A to regB
            
            push-a
            pop-b
            pop-a
            
            # Registers now:
                
                regA = 0
                regB = B
                regC = not(A)
            
            # nand them to c
            nand-c
            
            # swap to other stack
            
            swapstack
            
            # retrieve first result to regb
            
            pop-b
            
            # nand
            nand-a
            
            io-a
        
"""
world = World(1, replacement_strategy = "neighborhood", cm_prob = 0, ins_prob = 0, del_prob = 0, notify_ = True)

world.place_custom(program_xor)

world.schedule(50)


EMULATOR AT POSITION (0, 0) COMPUTED XOR
INPUT: [4180384015 2153942408]
OUTPUT: 2034832519


In [57]:
"""
Test EQU:
    
    An organism that computes EQU
    
    EQU(A,B) = (A and B) or (~A and ~B)
    
    not(nand(a,b)) or not(nand(not(a),not(b)))
    
    nand(nand(a,b),nand(not(a),not(b)))
    
    #0: Input a,b; store in active stack
    #1: Switch stack
    #2: Compute nand(a,b)
    #3: Clear register b
    #4: Store nand(a,b) to stack
    #5: Switch stack
    #6: Put a to reg_a, b to reg_c
    #7 : Calculate not(b) in reg_c, store to stack
    #8: Move a to reg_c by doing push-a pop-c
    #9: Calculate not(a) in reg_c
    #10: pop not(b) to reg_b
    #11: nand them to c
    #12: swap stack
    #13: pop to b
    #14: nand to a
    #15: io-a
        
"""
world = World(1, replacement_strategy = "neighborhood", cm_prob = 0, ins_prob = 0, del_prob = 0, notify_ = True)

world.place_custom(program_equ)

world.schedule(60)


EMULATOR AT POSITION (0, 0) COMPUTED EQU
INPUT: [3400156925  516246684]
OUTPUT: 730920350


In [6]:
"""
Test 10:
    
    Move
    
"""

program = [26,0,0,1,2]

world = World(3, replacement_strategy = "neighborhood", cm_prob = 0, ins_prob = 0, del_prob = 0)

world.place_custom(program,(1,1))

print("Before Moving")
print(world.pool.get())
print("\n")
print(world.rates)
print


world.schedule(1)

print("\nAfter Moving")
print(world.pool.get())
print("\n")
print(world.rates)

Before Moving
[[0 0 0]
 [0 <DigitalOrganism.CPUEmulator object at 0x000001F8D72DAC10> 0]
 [0 0 0]]


[[0 0 0]
 [0 5 0]
 [0 0 0]]

After Moving
[[0 0 0]
 [0 0 <DigitalOrganism.CPUEmulator object at 0x000001F8D72DAC10>]
 [0 0 0]]


[[0 0 0]
 [0 0 5]
 [0 0 0]]


In [15]:
"""
Test 11: 

    Sexual Reproduction
    
"""

program1 = [1,1,1,27,5,6,7,8]
program2 = [0,0,0,0,0,0]

world = World(3,cm_prob = 0, ins_prob = 0, del_prob = 0)

world.place_custom(program1, (0,0))
world.place_custom(program2, (0,1))

In [23]:
world.schedule(1)
print(world)

Emulator 1: 

Register A: 0
Register B: 0
Register C: 0
Instruction Pointer: 8
Memory: [1, 1, 1, 27, 5, 6, 7, 8]
Age: 8

Emulator 2: 

Register A: 0
Register B: 0
Register C: 0
Instruction Pointer: 2
Memory: [0, 0, 0, 0, 0, 0]
Age: 8




In [7]:
world.rates

array([[8, 6, 0],
       [7, 0, 0],
       [0, 0, 0]])

# Debugging:

### 1: Find out what those organisms of length 10 are

In [16]:
world = World(2)
world.place_custom([10, 20, 2, 0, 21, 2, 13, 9, 13, 7, 13, 3, 23, 17, 7, 3, 7, 15, 20, 17, 22, 9, 18, 2, 12, 14, 2, 1,\
            18, 15, 11, 16, 0, 13, 7, 9, 2, 18, 17, 20, 3, 19, 19, 25, 2, 0, 17, 21, 0, 1])

In [33]:
def min_length(world):
    
    len_ = 100000
    
    for i in range(world.pool.shape[0]):
        for j in range(world.pool.shape[1]):
            if world.pool.get((i,j)) == 0:
                pass
            else:
                if len(world.pool.get((i,j)).original_memory) < len_:
                    len_ = len(world.pool.get((i,j)).original_memory)
                    
    return len_

In [34]:
def max_length(world):
    
    len_ = -1
    
    for i in range(world.pool.shape[0]):
        for j in range(world.pool.shape[1]):
            if world.pool.get((i,j)) == 0:
                pass
            else:
                if len(world.pool.get((i,j)).original_memory) > len_:
                    len_ = len(world.pool.get((i,j)).original_memory)
                    
    return len_

In [18]:
min_length(world)

50

In [ ]:
while min_length(world) > 10:
    world.schedule(50)

In [22]:
min_length(world)

60

In [23]:
print(world.pool.get((1,1)).original_memory)

[20, 1, 17, 0, 16, 21, 2, 1, 23, 11, 19, 17, 22, 23, 9, 20, 6, 7, 11, 9, 2, 25, 25, 24, 14, 24, 18, 11, 3, 11, 25, 3, 12, 18, 24, 18, 19, 15, 25, 1, 21, 17, 19, 0, 19, 3, 19, 25, 23, 19, 20, 14, 23, 19, 25, 0, 17, 19, 22, 0, 1]


Track its lineage:

In [13]:
import numpy as np

with open('divlog2.log') as f:
    content = f.readlines()
    
con = np.array(content)

In [14]:
current = '[10, 20, 2, 0, 21, 2, 5, 0, 22, 6, 16, 12, 19, 20, 9, 17, 10, 3, 5, 20, 19, 23, 10, 14, 5, 3, 14, 3, 12, 21, 2, 8, 2]\n'

lineage = []
lineage.append(current)

found_start = False

loc_old = -1
loc_new = -1

while not found_start:
    # Find the last occurence of the current organism as a child
    # If all of the occurences of the current organism are as parent, we have found the start of the lineage
    
    loc_old = loc_new
    
    temp = 0
    location = np.where(con == current)[0][-1]
    
    for i in np.where(con == current)[0]:
        if con[i+1] == "\n" and i <= location:
            location = i
            temp = 1
            
    if temp == 0:
        found_start = True
        
    loc_new = location
    
    if loc_old == loc_new:
        found_start = True
    else:
        # Find the parent, append it to lineage
        lineage.append(con[location-1])
    
        # current is now the parent
        current = con[location-1]

lineage.reverse()

print("Full NAND lineage:\n")

for element in lineage:
    print(element)

Full NAND lineage:

[10, 20, 2, 0, 21, 2, 13, 9, 13, 7, 13, 3, 23, 17, 7, 3, 7, 15, 20, 17, 22, 9, 18, 2, 12, 14, 2, 1, 18, 15, 11, 16, 0, 13, 7, 9, 2, 18, 17, 20, 3, 19, 19, 25, 2, 0, 17, 21, 0, 1]

[10, 20, 2, 0, 21, 2, 13, 9, 13, 7, 13, 3, 23, 17, 7, 3, 7, 15, 20, 17, 22, 9, 18, 2, 12, 14, 2, 1, 18, 15, 11, 16, 0, 13, 7, 9, 2, 18, 17, 20, 3, 19, 19, 25, 2, 0, 17, 21, 0, 1]

[10, 20, 2, 0, 21, 2, 13, 9, 8, 7, 13, 3, 23, 17, 3, 7, 15, 20, 17, 22, 9, 18, 2, 12, 14, 2, 1, 2, 11, 11, 16, 0, 13, 7, 17, 2, 18, 19, 20, 3, 19, 19, 25, 2, 0, 17, 21, 0, 1]

[10, 20, 2, 0, 21, 2, 13, 9, 13, 7, 12, 3, 23, 17, 3, 9, 15, 20, 17, 22, 9, 18, 2, 12, 14, 2, 1, 4, 11, 11, 16, 0, 3, 13, 7, 17, 2, 18, 19, 20, 3, 19, 19, 25, 2, 0, 17, 21, 0, 1]

[10, 20, 2, 0, 21, 2, 15, 9, 13, 7, 12, 3, 23, 17, 9, 9, 15, 20, 17, 22, 3, 18, 2, 12, 14, 2, 1, 4, 11, 11, 16, 16, 0, 3, 13, 7, 17, 2, 18, 19, 20, 3, 19, 19, 25, 2, 0, 17, 21, 0, 1]

[10, 20, 2, 0, 21, 2, 15, 9, 3, 7, 12, 3, 23, 17, 9, 9, 15, 20, 17, 22, 3, 18, 2

Parsing mutations in the new lineage tracking system:



In [14]:
start = [16, 20, 2, 0, 21, 2, 20, 19, 25, 2, 0, 17, 21, 0, 1]
mutations = [['D', 2]]

In [15]:
for mutation in mutations:
    
    if mutation[0] == "C":
        start[mutation[1]] = mutation[2]
        
    if mutation[0] == "I":
        start.insert(mutation[1], mutation[2])
        
    if mutation[0] == "D":
        del start[mutation[1]]

# EXPERIMENTS:

### Default -> Not, run 1:


- Parameters: ins_prob = 0.05, del_prob = 0.05, cm_prob = 0.0025, N=30
- Instruction Set: Default
- Runtime: 10min 58s
- Program: [16, 15, 22, 0, 21, 2, 2, 2, 2, 2, 19, 9, 2, 12, 2, 2, 4, 4, 2, 2, 2, 17, 2, 2, 2, 21, 1, 2, 2, 17, 6, 2, 18, 2, 17, 2, 4, 8, 2, 2, 2, 15, 18, 20, 19, 19, 19, 25, 2, 0, 17, 21, 0, 1]
- Program Length: 54
- Number of Mutations: 46, of which:
-   24 Copy
-   11 Insertion
-   11 Deletion

#### Interesting Remark: Even though there was an equal number of insertion and deletion mutations, likely due to a change in the flow of the program, the end organism ended up being longer than the original organism (length 50)

All Mutations:

[['D', 18],
 ['C', 12, 8],
 ['I', 41, 19],
 ['C', 25, 9],
 ['C', 2, 22],
 ['D', 28],
 ['C', 28, 10],
 ['I', 40, 19],
 ['C', 1, 19],
 ['C', 32, 20],
 ['I', 23, 22],
 ['C', 31, 18],
 ['I', 16, 4],
 ['C', 26, 18],
 ['D', 17],
 ['D', 34],
 ['C', 41, 11],
 ['D', 11],
 ['D', 24],
 ['I', 17, 4],
 ['C', 34, 5],
 ['D', 18],
 ['C', 38, 15],
 ['I', 36, 18],
 ['C', 34, 17],
 ['D', 44],
 ['D', 7],
 ['I', 36, 5],
 ['C', 35, 4],
 ['D', 32],
 ['C', 25, 1],
 ['I', 39, 18],
 ['C', 10, 19],
 ['C', 11, 9],
 ['I', 28, 17],
 ['C', 29, 6],
 ['I', 13, 12],
 ['C', 37, 8],
 ['I', 41, 15],
 ['C', 47, 2],
 ['D', 43],
 ['C', 21, 17],
 ['C', 40, 2],
 ['C', 25, 21],
 ['C', 5, 2],
 ['C', 1, 15]]
 
 Mutations to be read as: [TYPE, LOCATION, INSTRUCTION]
 
 For Deletion Mutations the "INSTRUCTION" element is omitted

### Default -> Not, run 2:


- Parameters: ins_prob = 0.05, del_prob = 0.05, cm_prob = 0.05, N=30
- Instruction Set: Default
- Runtimes: 56s
- Program Length: 51
- Number of Mutations: 64, of which:
-   61 Copy
-   2 Insertion
-   1 Deletion

#### NOTE: With a higher copy probability the runtime is reduced by a factor of 10. This is subject to randomness.

### Default -> Not, run 3:


- Parameters: ins_prob = 0.05, del_prob = 0.05, cm_prob = 0.05, N=30
- Instruction Set: Default
- Runtimes: 36s
- Program Length: 51
- Number of Mutations: 64, of which:
-   61 Copy
-   2 Insertion
-   1 Deletion

#### NOTE: With a higher copy probability the runtime is reduced by a factor of 10. This is subject to randomness.

### Default to NOT, run 4:

- Parameters: ins_prob = 0.05, del_prob = 0.05, cm_prob = 0.05, N=30
- Instruction Set: Default
- Runtime: 1min 6s
- Program Length: 49
- Number of Mutations: 23, of which:
-   22 Copy
-   0 Insertion
-   1 Deletion


### NOT -> OR:

- Parameters: ins_prob = 0.05, del_prob = 0.05, cm_prob = 0.05, N=30
- Instruction Set: Default
- Runtime: 15min 31s
- Program Length: 52
- Number of Mutations: 212, of which:
-   202 Copy
-   5 Insertion
-   5 Deletion

### Default -> NOR:

- Parameters: ins_prob = 0.05, del_prob = 0.05, cm_prob = 0.05, N=30
- Instruction Set: Default
- Runtime: 27min 11s
- Program Length: 44
- Number of Mutations:
- Copy: 738
- Insert: 15
- Delete: 21
- TOTAL: 774

Only computes NOR

###

23min for NOT -> NOR

0.05 ins
0.05 del
0.01 cm
custom IS

# RUN EXPERIMENT HERE:

In [25]:
%%time

# Define the size of the pool here:
N = 30

# Define the start, targ£et and parameters of the experiment

# stat_cycles determines how often you're informed of the statistics of the population

experiment = Experiment(start_organism="default", target_function="nor", ins_prob = 0.05,\
                        del_prob = 0.05, cm_prob = 0.005,N=N, stat_cycles = 1000, instruction_set = "custom")

Wall time: 62.5 ms


In [ ]:
%%time

# Runs the experiment and looks for the target function
experiment.run()

# Captures the emulator which computes the functions realiably for further analysis
em = None

counter = 0

for i in range(N):
    for j in range(N):
        if experiment.world.pool.get((i,j)).original_memory == experiment.first_specimen:
            counter += 1
            em = experiment.world.pool.get((i,j))



Min Length: (46, (0, 2))
Max Length: (52, (0, 15))
Mean Length: 50.08
Min Rate: 46
Max Rate: 52
Mean Rate: 50.08
Max Age: 518
Mean Age: 131.27333333333334
NOT: 0
NAND: 0
AND: 0
OR_N: 0
OR: 0
AND_N: 0
NOR: 0
XOR: 0
EQU: 0


Min Length: (45, (4, 18))
Max Length: (53, (17, 7))
Mean Length: 50.044444444444444
Min Rate: 45
Max Rate: 53
Mean Rate: 50.044444444444444
Max Age: 566
Mean Age: 185.81222222222223
NOT: 0
NAND: 0
AND: 0
OR_N: 0
OR: 0
AND_N: 0
NOR: 0
XOR: 0
EQU: 0


Min Length: (46, (0, 22))
Max Length: (53, (1, 16))
Mean Length: 50.16111111111111
Min Rate: 46
Max Rate: 53
Mean Rate: 50.16111111111111
Max Age: 590
Mean Age: 215.98777777777778
NOT: 0
NAND: 0
AND: 0
OR_N: 0
OR: 0
AND_N: 0
NOR: 0
XOR: 0
EQU: 0


Min Length: (45, (1, 25))
Max Length: (54, (4, 13))
Mean Length: 50.12111111111111
Min Rate: 45
Max Rate: 54
Mean Rate: 50.12111111111111
Max Age: 629
Mean Age: 245.62333333333333
NOT: 0
NAND: 0
AND: 0
OR_N: 0
OR: 0
AND_N: 0
NOR: 0
XOR: 0
EQU: 0


Min Length: (42, (29, 25))
Ma



Min Length: (49, (14, 0))
Max Length: (60, (25, 19))
Mean Length: 55.37
Min Rate: 49
Max Rate: 60
Mean Rate: 55.37
Max Age: 294
Mean Age: 102.80444444444444
NOT: 0
NAND: 0
AND: 0
OR_N: 0
OR: 0
AND_N: 0
NOR: 0
XOR: 0
EQU: 0
Reacting on nor


Min Length: (50, (7, 11))
Max Length: (61, (27, 16))
Mean Length: 55.68222222222222
Min Rate: 50
Max Rate: 928
Mean Rate: 66.39222222222222
Max Age: 1801
Mean Age: 105.69555555555556
NOT: 0
NAND: 0
AND: 0
OR_N: 0
OR: 0
AND_N: 0
NOR: 1
XOR: 0
EQU: 0


Min Length: (50, (12, 12))
Max Length: (62, (19, 16))
Mean Length: 56.09444444444444
Min Rate: 50
Max Rate: 62
Mean Rate: 56.09444444444444
Max Age: 293
Mean Age: 107.06555555555556
NOT: 0
NAND: 0
AND: 0
OR_N: 0
OR: 0
AND_N: 0
NOR: 0
XOR: 0
EQU: 0


Min Length: (50, (3, 7))
Max Length: (61, (6, 26))
Mean Length: 56.276666666666664
Min Rate: 50
Max Rate: 61
Mean Rate: 56.276666666666664
Max Age: 304
Mean Age: 104.97222222222223
NOT: 0
NAND: 0
AND: 0
OR_N: 0
OR: 0
AND_N: 0
NOR: 0
XOR: 0
EQU: 0


Min Len

Notifying experiment of function_IO
Reacting on IO
1102531124 3043733047 4127190583
Notifying experiment of function_IO
Reacting on IO
2786277458 3957776194 4026007378
Notifying experiment of function_IO
Reacting on IO
2786277458 3957776194 68231184
Notifying experiment of function_IO
Reacting on IO
2795385077 1346787987 4141775607
Notifying experiment of function_IO
Reacting on IO
2962507749 3615163142 4160702439
Notifying experiment of function_IO
Reacting on IO
2962507749 3615163142 545539297
Notifying experiment of function_IO
Reacting on IO
4158549560 1854061832 4292804408
Notifying experiment of function_IO
Reacting on IO
3414450712 646387218 4018632218
Notifying experiment of function_IO
Reacting on IO
3414450712 646387218 3372245000
Notifying experiment of function_IO
Reacting on IO
111761283 638213349 648765415
Notifying experiment of function_IO
Reacting on IO
3422371417 73399668 3489660797
Notifying experiment of function_IO
Reacting on IO
3422371417 73399668 3416261129
Noti

Notifying experiment of function_IO
Reacting on IO
2405722898 538914380 2944364382
Notifying experiment of function_IO
Reacting on IO
2004771107 1882604841 2004772139
Notifying experiment of function_IO
Reacting on IO
92049616 1767667494 69238992
Notifying experiment of function_IO
Reacting on IO
2396640241 3064122390 3204177911
Notifying experiment of function_IO
Reacting on IO
2395632108 1235442134 2252873768
Notifying experiment of function_IO
Reacting on IO
3458475022 319966482 3744886046
Notifying experiment of function_IO
Reacting on IO
2595917350 2104054960 2190641670
Notifying experiment of function_IO
Reacting on IO
1592611518 3963417391 4278055871
Notifying experiment of function_IO
Reacting on IO
551996734 1274349574 1811402046
Notifying experiment of function_IO
Reacting on IO
3929087572 3452214694 573637200
Notifying experiment of function_IO
Reacting on IO
3316942552 2727790848 3887517656
Notifying experiment of function_IO
Reacting on IO
4260046660 925548433 3368091716
N

Notifying experiment of function_IO
Reacting on IO
2715988169 2614735996 3153835261
Notifying experiment of function_IO
Reacting on IO
2289492246 2506517702 2641817046
Notifying experiment of function_IO
Reacting on IO
2289492246 2506517702 135299344
Notifying experiment of function_IO
Reacting on IO
2956391731 3950377048 4218944891
Notifying experiment of function_IO
Reacting on IO
3213520797 363661499 3215945663
Notifying experiment of function_IO
Reacting on IO
3213520797 363661499 2852284164
Notifying experiment of function_IO
Reacting on IO
2505988485 1097287993 3581893053
Notifying experiment of function_IO
Reacting on IO
1641159950 2266607456 3889938286
Notifying experiment of function_IO
Reacting on IO
1641159950 2266607456 1623330830
Notifying experiment of function_IO
Reacting on IO
2292818785 1266177716 3422156789
Notifying experiment of function_IO
Reacting on IO
665733654 1718776381 1744730687
Notifying experiment of function_IO
Reacting on IO
665733654 1718776381 25954306



Min Length: (63, (0, 26))
Max Length: (72, (8, 8))
Mean Length: 68.45111111111112
Min Rate: 63
Max Rate: 557056
Mean Rate: 7515.648888888889
Max Age: 1632
Mean Age: 137.8
NOT: 424
NAND: 0
AND: 0
OR_N: 3
OR: 416
AND_N: 295
NOR: 0
XOR: 0
EQU: 0
Reacting on nor
Notifying experiment of function_IO
Reacting on IO
897756554 257751810 3397210741
Notifying experiment of function_IO
Reacting on IO
3000802877 1901485185 4091469501
Notifying experiment of function_IO
Reacting on IO
1332672219 1594833608 1601175259
Notifying experiment of function_IO
Reacting on IO
1332672219 1594833608 6341651
Notifying experiment of function_IO
Reacting on IO
3280925540 3736016241 1014041755
Notifying experiment of function_IO
Reacting on IO
4251350401 1640694257 4260328945
Notifying experiment of function_IO
Reacting on IO
669139899 127443017 670753787
Notifying experiment of function_IO
Reacting on IO
669139899 127443017 543310770
Notifying experiment of function_IO
Reacting on IO
845477439 1541554796 344948

Notifying experiment of function_IO
Reacting on IO
368339777 4070915572 3926627518
Notifying experiment of function_IO
Reacting on IO
2472181493 2508608718 2547941119
Notifying experiment of function_IO
Reacting on IO
3694344623 3178089534 4253023679
Notifying experiment of function_IO
Reacting on IO
3694344623 3178089534 1074934145
Notifying experiment of function_IO
Reacting on IO
1143138944 3015431933 3151828351
Notifying experiment of function_IO
Reacting on IO
2302235987 346326530 2646464339
Notifying experiment of function_IO
Reacting on IO
1256916991 3580491743 3756783615
Notifying experiment of function_IO
Reacting on IO
1256916991 3580491743 176291872
Notifying experiment of function_IO
Reacting on IO
1844589142 2752547692 2450378153
Notifying experiment of function_IO
Reacting on IO
1037848923 777508401 1071634299
Notifying experiment of function_IO
Reacting on IO
3231940896 2083884167 4239887783
Notifying experiment of function_IO
Reacting on IO
3231940896 2083884167 2156003

Notifying experiment of function_IO
Reacting on IO
4133851564 1151811631 4142387631
Notifying experiment of function_IO
Reacting on IO
2514300771 2115092225 4292722531
Notifying experiment of function_IO
Reacting on IO
2514300771 2115092225 2177630306
Notifying experiment of function_IO
Notifying experiment of function_IO
Reacting on IO
3661909243 2518453939 3730624251
Notifying experiment of function_IO
Reacting on IO
1722175504 156793457 1878948465
Notifying experiment of function_IO
Reacting on IO
1722175504 156793457 1722155008
Notifying experiment of function_IO
Notifying experiment of function_IO
Reacting on IO
568975426 1333008992 1878915682
Notifying experiment of function_IO
Reacting on IO
2921012809 3225931042 3999260523
Notifying experiment of function_IO
Reacting on IO
2921012809 3225931042 773329481
Notifying experiment of function_IO
Notifying experiment of function_IO
Reacting on IO
4058553120 1778639465 4226546537
Notifying experiment of function_IO
Reacting on IO
55938

Reacting on nor
Notifying experiment of function_IO
Reacting on IO
2180850895 3785527322 3791631583
Notifying experiment of function_IO
Reacting on IO
201856426 4092224989 4293555711
Notifying experiment of function_IO
Reacting on IO
201856426 4092224989 201330722
Notifying experiment of function_IO
Reacting on IO
518686957 1091353130 3776280338
Notifying experiment of function_IO
Reacting on IO
3717931941 2143628872 4292816877
Notifying experiment of function_IO
Reacting on IO
3717931941 2143628872 2149188005
Notifying experiment of function_IO
Reacting on IO
3162279107 3240137978 4252802299
Notifying experiment of function_IO
Reacting on IO
285620556 3574649325 3575052781
Notifying experiment of function_IO
Reacting on IO
285620556 3574649325 403456
Notifying experiment of function_IO
Reacting on IO
4199366522 6906549 95600773
Notifying experiment of function_IO
Reacting on IO
1310761494 268656808 1579410110
Notifying experiment of function_IO
Reacting on IO
1310761494 268656808 1310

Notifying experiment of function_IO
Reacting on IO
3151898121 2902924177 3219079065
Notifying experiment of function_IO
Reacting on IO
2322158576 950540851 3135855603
Notifying experiment of function_IO
Reacting on IO
2322158576 950540851 2185314752
Notifying experiment of function_IO
Reacting on IO
1104300484 2908872457 3190666811
Notifying experiment of function_IO
Reacting on IO
4210625307 2467067690 4227694395
Notifying experiment of function_IO
Reacting on IO
4210625307 2467067690 1760626705
Reacting on nor
Reacting on nor
Notifying experiment of function_IO
Reacting on IO
4039651454 526537629 4293555199
Notifying experiment of function_IO
Reacting on IO
283796251 1961371860 1961520095
Notifying experiment of function_IO
Reacting on IO
283796251 1961371860 148235
Notifying experiment of function_IO
Reacting on IO
1650822903 484520873 2644144392
Notifying experiment of function_IO
Reacting on IO
3610581495 4235106945 4286553079
Notifying experiment of function_IO
Reacting on IO
361

Notifying experiment of function_IO
Reacting on IO
1946613981 2781305522 2348353314
Notifying experiment of function_IO
Reacting on IO
274441942 1065160744 1065334526
Notifying experiment of function_IO
Reacting on IO
274441942 1065160744 173782
Notifying experiment of function_IO
Reacting on IO
1421185392 188947853 1610063357
Notifying experiment of function_IO
Reacting on IO
2671332925 2039228719 4290604863
Notifying experiment of function_IO
Reacting on IO
2671332925 2039228719 2251376144
Notifying experiment of function_IO
Reacting on IO
2349526650 358540165 1945440645
Notifying experiment of function_IO
Reacting on IO
4148898460 2728814203 4159663871
Notifying experiment of function_IO
Reacting on IO
4148898460 2728814203 1430849668
Reacting on nor
Notifying experiment of function_IO
Reacting on IO
2531715499 227275194 2683304379
Notifying experiment of function_IO
Reacting on IO
264948873 152678567 266075311
Notifying experiment of function_IO
Reacting on IO
264948873 152678567 1

Notifying experiment of function_IO
Reacting on IO
233136078 48558597 266731471
Notifying experiment of function_IO
Reacting on IO
3564821635 845469647 4135509967
Notifying experiment of function_IO
Reacting on IO
3564821635 845469647 3290040320
Notifying experiment of function_IO
Reacting on IO
3077951360 1745803341 1217015935
Notifying experiment of function_IO
Reacting on IO
140683098 682687461 687013887
Notifying experiment of function_IO
Reacting on IO
140683098 682687461 4326426
Notifying experiment of function_IO
Reacting on IO
1647987368 2480937490 4093270714
Notifying experiment of function_IO
Reacting on IO
3760502758 2405092240 4018077686
Notifying experiment of function_IO
Reacting on IO
3760502758 2405092240 1612985446
Notifying experiment of function_IO
Reacting on IO
984201293 4256208078 3310766002
Notifying experiment of function_IO
Reacting on IO
2438695937 883501921 3053170529
Notifying experiment of function_IO
Reacting on IO
2438695937 883501921 2169668608
Notifying

Reacting on nor
Notifying experiment of function_IO
Reacting on IO
2714265078 3369509030 3923699190
Notifying experiment of function_IO
Reacting on IO
204667674 3802113460 4005560254
Notifying experiment of function_IO
Reacting on IO
204667674 3802113460 203446794
Notifying experiment of function_IO
Reacting on IO
4002789699 340629638 292177596
Notifying experiment of function_IO
Reacting on IO
2023630093 568284263 2044687727
Notifying experiment of function_IO
Reacting on IO
2023630093 568284263 1476403464
Notifying experiment of function_IO
Reacting on IO
751116056 2804730018 2951602106
Notifying experiment of function_IO
Reacting on IO
883667339 2381792923 3187668891
Notifying experiment of function_IO
Reacting on IO
883667339 2381792923 805875968
Notifying experiment of function_IO
Reacting on IO
4135958967 4278867186 159008328
Notifying experiment of function_IO
Reacting on IO
1622851459 2968150965 4042981303
Notifying experiment of function_IO
Reacting on IO
1622851459 2968150965

Reacting on nor
Notifying experiment of function_IO
Reacting on IO
3631785526 1473656019 3758012151
Notifying experiment of function_IO
Reacting on IO
2583438570 958385697 3120557803
Notifying experiment of function_IO
Reacting on IO
2583438570 958385697 2162172106
Notifying experiment of function_IO
Reacting on IO
3631523225 1419908652 3707156413
Notifying experiment of function_IO
Reacting on IO
676090217 1557434425 2094846329
Notifying experiment of function_IO
Reacting on IO
676090217 1557434425 537411904
Notifying experiment of function_IO
Reacting on IO
3440399479 164263814 3453646839
Notifying experiment of function_IO
Reacting on IO
2478862302 671125111 3149983743
Notifying experiment of function_IO
Reacting on IO
2478862302 671125111 2478858632
Notifying experiment of function_IO
Reacting on IO
230600364 29274377 230604717
Notifying experiment of function_IO
Reacting on IO
1907547712 2879013873 4223328241
Notifying experiment of function_IO
Reacting on IO
1907547712 2879013873

Notifying experiment of function_IO
Reacting on IO
1089327017 3576245776 3589140409
Notifying experiment of function_IO
Reacting on IO
509665185 3484197050 3756851131
Notifying experiment of function_IO
Reacting on IO
509665185 3484197050 272654081
Notifying experiment of function_IO
Reacting on IO
9922995 690721882 4285044300
Notifying experiment of function_IO
Reacting on IO
1624616326 1624850404 1625145318
Notifying experiment of function_IO
Reacting on IO
1624616326 1624850404 294914
Notifying experiment of function_IO
Reacting on IO
3704960749 2814357832 4294948845
Notifying experiment of function_IO
Reacting on IO
3022596392 2942566287 3211526063
Notifying experiment of function_IO
Reacting on IO
3022596392 2942566287 268959776
Notifying experiment of function_IO
Reacting on IO
2789875158 1997487370 1505092137
Notifying experiment of function_IO
Reacting on IO
2979723884 2151404432 2981886972
Notifying experiment of function_IO
Reacting on IO
2979723884 2151404432 830482540
Notif



Min Length: (67, (3, 15))
Max Length: (74, (0, 26))
Mean Length: 71.57777777777778
Min Rate: 68
Max Rate: 4653056
Mean Rate: 27326.816666666666
Max Age: 1145
Mean Age: 138.5311111111111
NOT: 138
NAND: 10
AND: 2
OR_N: 485
OR: 390
AND_N: 244
NOR: 0
XOR: 0
EQU: 0
Reacting on nor
Notifying experiment of function_IO
Reacting on IO
3368897532 1206427621 3488473085
Notifying experiment of function_IO
Reacting on IO
2503434812 784083299 3216993151
Notifying experiment of function_IO
Reacting on IO
2503434812 784083299 2432909852
Notifying experiment of function_IO
Reacting on IO
3920623451 745389278 3991928799
Notifying experiment of function_IO
Reacting on IO
724178355 3090280609 3141268915
Notifying experiment of function_IO
Reacting on IO
724178355 3090280609 50988306
Notifying experiment of function_IO
Reacting on IO
1706178272 245515298 1873966818
Notifying experiment of function_IO
Reacting on IO
2363927432 806229005 3169763213
Notifying experiment of function_IO
Reacting on IO
2363927

Notifying experiment of function_IO
Reacting on IO
2706766089 720740532 2885024189
Notifying experiment of function_IO
Reacting on IO
3056026491 1072229905 3220172667
Notifying experiment of function_IO
Reacting on IO
3056026491 1072229905 2147942762
Notifying experiment of function_IO
Reacting on IO
962988559 3187649855 3331978736
Notifying experiment of function_IO
Reacting on IO
2798828557 1607253055 4292787263
Notifying experiment of function_IO
Reacting on IO
2798828557 1607253055 2685534208
Notifying experiment of function_IO
Reacting on IO
1976196365 453979978 2144304975
Notifying experiment of function_IO
Reacting on IO
2134972296 3343738881 4283269001
Notifying experiment of function_IO
Reacting on IO
2134972296 3343738881 939530120
Notifying experiment of function_IO
Reacting on IO
410235294 2073152207 3884732001
Notifying experiment of function_IO
Reacting on IO
1262114867 4218673894 4219198199
Notifying experiment of function_IO
Reacting on IO
1262114867 4218673894 524305
N

Notifying experiment of function_IO
Reacting on IO
4263562876 3908269915 4277403519
Notifying experiment of function_IO
Reacting on IO
1476707061 1763403939 2032134903
Notifying experiment of function_IO
Reacting on IO
1476707061 1763403939 268730964
Notifying experiment of function_IO
Reacting on IO
3980126393 639075524 314840902
Notifying experiment of function_IO
Reacting on IO
1633307898 4199386053 4217360383
Notifying experiment of function_IO
Reacting on IO
1633307898 4199386053 17974330
Notifying experiment of function_IO
Reacting on IO
3918103759 3707499088 4261148383
Notifying experiment of function_IO
Reacting on IO
143604484 1803802884 1804598020
Notifying experiment of function_IO
Reacting on IO
143604484 1803802884 795136
Notifying experiment of function_IO
Reacting on IO
2672852452 499890264 1622114843
Notifying experiment of function_IO
Reacting on IO
3314345035 2831066351 3988710639
Notifying experiment of function_IO
Reacting on IO
3314345035 2831066351 1157644288
Noti

Reacting on nor


Min Length: (68, (28, 2))
Max Length: (75, (6, 26))
Mean Length: 72.05888888888889
Min Rate: 68
Max Rate: 4653056
Mean Rate: 26456.442222222224
Max Age: 1295
Mean Age: 147.73777777777778
NOT: 153
NAND: 9
AND: 0
OR_N: 502
OR: 399
AND_N: 232
NOR: 0
XOR: 0
EQU: 0
Reacting on nor
Reacting on nor
Notifying experiment of function_IO
Reacting on IO
3299396682 2437401238 3589263070
Notifying experiment of function_IO
Reacting on IO
690804990 1827000946 1844304638
Notifying experiment of function_IO
Reacting on IO
690804990 1827000946 17303692
Notifying experiment of function_IO
Reacting on IO
3403196946 2421345868 891770349
Notifying experiment of function_IO
Reacting on IO
1370752783 1300884717 1572728815
Notifying experiment of function_IO
Reacting on IO
1370752783 1300884717 271844098
Notifying experiment of function_IO
Reacting on IO
1106877530 3904581711 3925589087
Notifying experiment of function_IO
Reacting on IO
1855249471 4165581427 4275961471
Notifying experiment of

Notifying experiment of function_IO
Reacting on IO
1616244599 1704458859 1708654463
Notifying experiment of function_IO
Reacting on IO
572889205 341063474 913686391
Notifying experiment of function_IO
Reacting on IO
572889205 341063474 572622917
Notifying experiment of function_IO
Reacting on IO
2819610723 3103381886 1475356572
Notifying experiment of function_IO
Reacting on IO
2443611142 1988613072 4154974166
Notifying experiment of function_IO
Reacting on IO
2443611142 1988613072 2166361094
Notifying experiment of function_IO
Reacting on IO
2816369329 2339365866 2952782843
Notifying experiment of function_IO
Reacting on IO
617728405 1323539608 1861475741
Notifying experiment of function_IO
Reacting on IO
617728405 1323539608 537936133
Notifying experiment of function_IO
Reacting on IO
1003695258 3061482444 3291272037
Notifying experiment of function_IO
Reacting on IO
1447029741 1125311519 1463807999
Notifying experiment of function_IO
Reacting on IO
1447029741 1125311519 338496480
No



Min Length: (68, (28, 2))
Max Length: (76, (2, 18))
Mean Length: 73.0011135857461
Min Rate: 0
Max Rate: 4915200
Mean Rate: 27230.86111111111
Max Age: 2148
Mean Age: 135.22
NOT: 135
NAND: 8
AND: 1
OR_N: 467
OR: 378
AND_N: 221
NOR: 0
XOR: 0
EQU: 0
Reacting on nor
Notifying experiment of function_IO
Reacting on IO
1800619457 3404295502 3959123407
Notifying experiment of function_IO
Reacting on IO
1043189825 2815495283 3221085299
Notifying experiment of function_IO
Reacting on IO
1043189825 2815495283 405590016
Notifying experiment of function_IO
Reacting on IO
2910287223 2700241284 1384680072
Notifying experiment of function_IO
Reacting on IO
2409515319 3193202654 3219025919
Notifying experiment of function_IO
Reacting on IO
2409515319 3193202654 25823265
Notifying experiment of function_IO
Reacting on IO
3127163404 2527672038 3203251950
Notifying experiment of function_IO
Reacting on IO
1406508151 904992463 2012585727
Notifying experiment of function_IO
Reacting on IO
1406508151 904992

Reacting on IO
232359789 1878115247 1878640623
Notifying experiment of function_IO
Reacting on IO
232359789 1878115247 525376
Notifying experiment of function_IO
Reacting on IO
1203151573 2964571767 3091815722
Notifying experiment of function_IO
Reacting on IO
198025952 3816902376 3956141800
Notifying experiment of function_IO
Reacting on IO
198025952 3816902376 139239424
Notifying experiment of function_IO
Reacting on IO
3513292486 1377320616 3547905774
Notifying experiment of function_IO
Reacting on IO
2730400525 510157287 3204381679
Notifying experiment of function_IO
Reacting on IO
2730400525 510157287 2694224392
Notifying experiment of function_IO
Reacting on IO
450911162 219820305 3844056133
Notifying experiment of function_IO
Reacting on IO
435948722 718791918 1006628094
Notifying experiment of function_IO
Reacting on IO
435948722 718791918 287836176
Notifying experiment of function_IO
Reacting on IO
1390557988 460643193 1542912893
Notifying experiment of function_IO
Reacting on

Reacting on nor
Notifying experiment of function_IO
Reacting on IO
2171147344 870105549 3019757021
Notifying experiment of function_IO
Reacting on IO
3320700055 2617397460 3723484375
Notifying experiment of function_IO
Reacting on IO
3320700055 2617397460 1106086915
Notifying experiment of function_IO
Reacting on IO
1271717178 2353904876 3023250117
Notifying experiment of function_IO
Reacting on IO
3957077952 3300903440 4026531792
Notifying experiment of function_IO
Reacting on IO
3957077952 3300903440 725628352
Notifying experiment of function_IO
Reacting on IO
4031579489 3454999710 4260306431
Notifying experiment of function_IO
Reacting on IO
3479963583 384672954 3756894143
Notifying experiment of function_IO
Reacting on IO
3479963583 384672954 3372221189
Notifying experiment of function_IO
Reacting on IO
3556319901 753335005 738647394
Notifying experiment of function_IO
Reacting on IO
3087139550 819779661 3101612767
Notifying experiment of function_IO
Reacting on IO
3087139550 81977

Reacting on nor
Notifying experiment of function_IO
Reacting on IO
3191249062 3079955709 3216440575
Notifying experiment of function_IO
Reacting on IO
4262666255 3571628499 4277385695
Notifying experiment of function_IO
Reacting on IO
4262666255 3571628499 705757196
Notifying experiment of function_IO
Reacting on IO
506243518 2536079004 3788723777
Notifying experiment of function_IO
Reacting on IO
2624696508 4199630084 4268881340
Notifying experiment of function_IO
Reacting on IO
2624696508 4199630084 69251256
Notifying experiment of function_IO
Reacting on IO
1697732643 2222088176 3849543667
Notifying experiment of function_IO
Reacting on IO
1405960237 2334088661 3688848893
Notifying experiment of function_IO
Reacting on IO
1405960237 2334088661 1354760232
Notifying experiment of function_IO
Reacting on IO
3491605239 794154930 803362056
Notifying experiment of function_IO
Reacting on IO
2055381783 904794143 2146349855
Notifying experiment of function_IO
Reacting on IO
2055381783 90479

Notifying experiment of function_IO
Reacting on IO
1780600107 161508860 1805774335
Notifying experiment of function_IO
Reacting on IO
3772200409 2255015529 3875536889
Notifying experiment of function_IO
Reacting on IO
3772200409 2255015529 1620521360
Notifying experiment of function_IO
Reacting on IO
1161495586 140924975 3133471709
Notifying experiment of function_IO
Reacting on IO
268738311 3148103856 3148144567
Notifying experiment of function_IO
Reacting on IO
268738311 3148103856 40711
Notifying experiment of function_IO
Reacting on IO
3163654043 3723071630 4260992927
Notifying experiment of function_IO
Reacting on IO
152194836 2006905603 2141126423
Notifying experiment of function_IO
Reacting on IO
152194836 2006905603 134220820
Notifying experiment of function_IO
Reacting on IO
696693133 2930520748 3598274162
Notifying experiment of function_IO
Reacting on IO
466874395 437549829 467139359
Notifying experiment of function_IO
Reacting on IO
466874395 437549829 29589530
Notifying ex

Notifying experiment of function_IO
Reacting on IO
1473046836 3090747359 4294832127
Notifying experiment of function_IO
Reacting on IO
1473046836 3090747359 1204084768
Notifying experiment of function_IO
Reacting on IO
1770910884 3308699447 3988783031
Notifying experiment of function_IO
Reacting on IO
63584575 3557962246 3621403455
Notifying experiment of function_IO
Reacting on IO
63584575 3557962246 63441209
Notifying experiment of function_IO
Reacting on IO
2091315195 925922578 2142699515
Notifying experiment of function_IO
Reacting on IO
1137426906 1494870787 1541139419
Notifying experiment of function_IO
Reacting on IO
1137426906 1494870787 46268632
Notifying experiment of function_IO
Reacting on IO
3823071132 4288475318 4292870078
Notifying experiment of function_IO
Reacting on IO
2541230901 1537888980 3757726709
Notifying experiment of function_IO
Reacting on IO
2541230901 1537888980 2219837729
Notifying experiment of function_IO
Reacting on IO
2146269962 1154060700 2146302878
N

Notifying experiment of function_IO
Reacting on IO
3709468051 1596552269 3745119711
Notifying experiment of function_IO
Reacting on IO
3978487828 1217143514 3987663582
Notifying experiment of function_IO
Reacting on IO
3978487828 1217143514 2770520068
Notifying experiment of function_IO
Reacting on IO
708338651 3631695575 3586628644
Notifying experiment of function_IO
Reacting on IO
1856683674 1001618301 2143027199
Notifying experiment of function_IO
Reacting on IO
1856683674 1001618301 1141408898
Notifying experiment of function_IO
Reacting on IO
802084747 668547728 803209115
Notifying experiment of function_IO
Reacting on IO
3797557395 2486122630 4135532695
Notifying experiment of function_IO
Reacting on IO
3797557395 2486122630 1649410065
Notifying experiment of function_IO
Reacting on IO
1429161545 3095598782 2865805750
Notifying experiment of function_IO
Reacting on IO
3417381964 3597665156 3757121484
Notifying experiment of function_IO
Reacting on IO
3417381964 3597665156 1594563



Min Length: (72, (1, 27))
Max Length: (78, (5, 3))
Mean Length: 75.80713489409142
Min Rate: 0
Max Rate: 4980736
Mean Rate: 27727.674444444445
Max Age: 2099
Mean Age: 130.15222222222224
NOT: 109
NAND: 9
AND: 1
OR_N: 448
OR: 330
AND_N: 194
NOR: 0
XOR: 0
EQU: 0
Reacting on nor
Notifying experiment of function_IO
Reacting on IO
2113707382 4098143184 4261199862
Notifying experiment of function_IO
Reacting on IO
1218826280 2799965384 4008039656
Notifying experiment of function_IO
Reacting on IO
1218826280 2799965384 1208074272
Notifying experiment of function_IO
Reacting on IO
3543449838 4063653667 751517457
Notifying experiment of function_IO
Reacting on IO
2324397095 1592932986 3741023871
Notifying experiment of function_IO
Reacting on IO
2324397095 1592932986 2148090885
Notifying experiment of function_IO
Reacting on IO
2349722002 4244132423 4244503511
Notifying experiment of function_IO
Reacting on IO
2284195495 3502661432 3638976447
Notifying experiment of function_IO
Reacting on IO
2



Min Length: (72, (25, 3))
Max Length: (78, (6, 10))
Mean Length: 75.66333333333333
Min Rate: 72
Max Rate: 4980736
Mean Rate: 28559.15111111111
Max Age: 1408
Mean Age: 153.2411111111111
NOT: 141
NAND: 6
AND: 0
OR_N: 544
OR: 405
AND_N: 247
NOR: 0
XOR: 0
EQU: 0
Reacting on nor
Notifying experiment of function_IO
Reacting on IO
2160867211 2993893247 3003072511
Notifying experiment of function_IO
Reacting on IO
4077914678 2929243132 4288199678
Notifying experiment of function_IO
Reacting on IO
4077914678 2929243132 1358956546
Notifying experiment of function_IO
Reacting on IO
2084491047 151720375 2210476248
Notifying experiment of function_IO
Reacting on IO
2983894670 3669841340 4227850174
Notifying experiment of function_IO
Reacting on IO
2983894670 3669841340 558008834
Notifying experiment of function_IO
Reacting on IO
2992515165 3001521634 3003104767
Notifying experiment of function_IO
Reacting on IO
3539692825 973072957 4227854141
Notifying experiment of function_IO
Reacting on IO
353

Notifying experiment of function_IO
Reacting on IO
3304451818 3514913750 3589668862
Notifying experiment of function_IO
Reacting on IO
4161682325 1143145066 4230953983
Notifying experiment of function_IO
Reacting on IO
4161682325 1143145066 3087808917
Notifying experiment of function_IO
Reacting on IO
729446570 487703136 3565520725
Notifying experiment of function_IO
Reacting on IO
1229700415 449354560 1540079487
Notifying experiment of function_IO
Reacting on IO
1229700415 449354560 1090724927
Notifying experiment of function_IO
Reacting on IO
899811802 725493744 1069428730
Notifying experiment of function_IO
Reacting on IO
3570508763 2708151578 4126916571
Notifying experiment of function_IO
Reacting on IO
3570508763 2708151578 1418764993
Notifying experiment of function_IO
Reacting on IO
1845535688 980696632 2449431607
Notifying experiment of function_IO
Reacting on IO
2897099609 1919123273 4277108569
Notifying experiment of function_IO
Reacting on IO
2897099609 1919123273 2357985296



Min Length: (69, (4, 14))
Max Length: (78, (1, 19))
Mean Length: 76.24471635150167
Min Rate: 0
Max Rate: 5046272
Mean Rate: 29176.704444444444
Max Age: 794
Mean Age: 146.4088888888889
NOT: 146
NAND: 7
AND: 1
OR_N: 476
OR: 370
AND_N: 232
NOR: 0
XOR: 0
EQU: 0


Min Length: (70, (11, 13))
Max Length: (79, (16, 21))
Mean Length: 76.43095768374165
Min Rate: 0
Max Rate: 5046272
Mean Rate: 28327.67111111111
Max Age: 1881
Mean Age: 137.73333333333332
NOT: 100
NAND: 6
AND: 2
OR_N: 435
OR: 322
AND_N: 199
NOR: 0
XOR: 0
EQU: 0
Reacting on nor
Notifying experiment of function_IO
Reacting on IO
2006428158 4118904047 4153933311
Notifying experiment of function_IO
Reacting on IO
1956749034 4053113058 4122320618
Notifying experiment of function_IO
Reacting on IO
1956749034 4053113058 69207560
Notifying experiment of function_IO
Reacting on IO
3653635408 869875119 641331887
Notifying experiment of function_IO
Reacting on IO
1264574770 810098577 2069888947
Notifying experiment of function_IO
Reacting o

Notifying experiment of function_IO
Reacting on IO
181361177 126049273 265247737
Notifying experiment of function_IO
Reacting on IO
1444897953 1254386774 1591700727
Notifying experiment of function_IO
Reacting on IO
1444897953 1254386774 337313953
Notifying experiment of function_IO
Reacting on IO
68888325 2284732021 4226078970
Notifying experiment of function_IO
Reacting on IO
764179376 1733735578 1876883386
Notifying experiment of function_IO
Reacting on IO
764179376 1733735578 143147808
Notifying experiment of function_IO
Reacting on IO
2415787781 2640419110 2684223271
Notifying experiment of function_IO
Reacting on IO
2474146731 2777373484 3086711727
Notifying experiment of function_IO
Reacting on IO
2474146731 2777373484 309338243
Notifying experiment of function_IO
Reacting on IO
2893538576 1453216017 1401428719
Notifying experiment of function_IO
Reacting on IO
755946856 3600322529 4288602089
Notifying experiment of function_IO
Reacting on IO
755946856 3600322529 688279560
Notif



Min Length: (72, (2, 6))
Max Length: (79, (4, 26))
Mean Length: 77.17633928571429
Min Rate: 0
Max Rate: 5046272
Mean Rate: 25210.086666666666
Max Age: 2104
Mean Age: 118.78222222222222
NOT: 87
NAND: 2
AND: 0
OR_N: 388
OR: 259
AND_N: 155
NOR: 0
XOR: 0
EQU: 0


Min Length: (74, (1, 5))
Max Length: (79, (7, 24))
Mean Length: 77.02004454342985
Min Rate: 0
Max Rate: 2555904
Mean Rate: 26271.032222222224
Max Age: 1684
Mean Age: 134.05777777777777
NOT: 137
NAND: 6
AND: 1
OR_N: 446
OR: 351
AND_N: 230
NOR: 0
XOR: 0
EQU: 0


Min Length: (73, (10, 28))
Max Length: (79, (2, 15))
Mean Length: 76.96218020022246
Min Rate: 0
Max Rate: 5046272
Mean Rate: 29354.334444444445
Max Age: 2124
Mean Age: 130.89222222222222
NOT: 107
NAND: 2
AND: 0
OR_N: 435
OR: 337
AND_N: 200
NOR: 0
XOR: 0
EQU: 0


Min Length: (73, (22, 6))
Max Length: (84, (5, 11))
Mean Length: 76.876254180602
Min Rate: 0
Max Rate: 4915200
Mean Rate: 28846.02666666667
Max Age: 1028
Mean Age: 132.65
NOT: 107
NAND: 4
AND: 1
OR_N: 427
OR: 321
A

Notifying experiment of function_IO
Reacting on IO
380869244 3533073488 3602382460
Notifying experiment of function_IO
Reacting on IO
2886255462 727806654 2942956542
Notifying experiment of function_IO
Reacting on IO
2886255462 727806654 2215149888
Notifying experiment of function_IO
Reacting on IO
1446764715 1462579688 2848202580
Notifying experiment of function_IO
Reacting on IO
169494011 4012156814 4013877247
Notifying experiment of function_IO
Reacting on IO
169494011 4012156814 1720433
Notifying experiment of function_IO
Reacting on IO
699865369 2267628299 2948553499
Notifying experiment of function_IO
Reacting on IO
2234826334 1484593947 3716005727
Notifying experiment of function_IO
Reacting on IO
2234826334 1484593947 2231411780
Notifying experiment of function_IO
Reacting on IO
373915456 2345329322 3921051839
Notifying experiment of function_IO
Reacting on IO
528408648 1618503767 2139023455
Notifying experiment of function_IO
Reacting on IO
528408648 1618503767 520519688
Notif

Reacting on nor


Min Length: (72, (22, 4))
Max Length: (79, (0, 16))
Mean Length: 76.79310344827586
Min Rate: 0
Max Rate: 4915200
Mean Rate: 29092.15111111111
Max Age: 1602
Mean Age: 143.38111111111112
NOT: 132
NAND: 8
AND: 0
OR_N: 469
OR: 347
AND_N: 219
NOR: 1
XOR: 0
EQU: 0
Reacting on nor
Notifying experiment of function_IO
Reacting on IO
3219017746 1996997466 4292858714
Notifying experiment of function_IO
Reacting on IO
2749118721 2479811960 3017735545
Notifying experiment of function_IO
Reacting on IO
2749118721 2479811960 537923585
Notifying experiment of function_IO
Reacting on IO
1852894637 410781586 2442072658
Notifying experiment of function_IO
Reacting on IO
3775883512 1260708156 3945785852
Notifying experiment of function_IO
Reacting on IO
3775883512 1260708156 2685077696
Notifying experiment of function_IO
Reacting on IO
1341452554 3413276078 3489068462
Notifying experiment of function_IO
Reacting on IO
1939028232 1580873977 2142976505
Notifying experiment of function_IO
R



Min Length: (73, (24, 3))
Max Length: (79, (0, 22))
Mean Length: 76.9988864142539
Min Rate: 0
Max Rate: 5046272
Mean Rate: 26431.481111111112
Max Age: 850
Mean Age: 147.38888888888889
NOT: 128
NAND: 7
AND: 0
OR_N: 440
OR: 358
AND_N: 213
NOR: 0
XOR: 0
EQU: 0


Min Length: (73, (1, 0))
Max Length: (79, (0, 25))
Mean Length: 77.03111111111112
Min Rate: 73
Max Rate: 5046272
Mean Rate: 27762.981111111112
Max Age: 1148
Mean Age: 132.18
NOT: 108
NAND: 9
AND: 0
OR_N: 470
OR: 369
AND_N: 208
NOR: 0
XOR: 0
EQU: 0


Min Length: (73, (1, 1))
Max Length: (80, (3, 7))
Mean Length: 77.30367074527253
Min Rate: 0
Max Rate: 5046272
Mean Rate: 29774.737777777777
Max Age: 2031
Mean Age: 145.2277777777778
NOT: 136
NAND: 9
AND: 0
OR_N: 498
OR: 390
AND_N: 229
NOR: 0
XOR: 0
EQU: 0


Min Length: (73, (6, 0))
Max Length: (79, (0, 4))
Mean Length: 77.36677814938685
Min Rate: 0
Max Rate: 5111808
Mean Rate: 26168.744444444445
Max Age: 1366
Mean Age: 116.07888888888888
NOT: 93
NAND: 6
AND: 0
OR_N: 367
OR: 219
AND_

Notifying experiment of function_IO
Reacting on IO
2838092959 736505899 2884631743
Notifying experiment of function_IO
Reacting on IO
2613525598 3500838887 3689922559
Notifying experiment of function_IO
Reacting on IO
2613525598 3500838887 189083672
Notifying experiment of function_IO
Reacting on IO
596706141 2219229215 3698261154
Notifying experiment of function_IO
Reacting on IO
1764420384 672958918 1765731302
Notifying experiment of function_IO
Reacting on IO
1764420384 672958918 1092772384
Notifying experiment of function_IO
Reacting on IO
3637355710 2447592160 3656380158
Notifying experiment of function_IO
Reacting on IO
1673230358 3397863822 3955194270
Notifying experiment of function_IO
Reacting on IO
1673230358 3397863822 557330448
Notifying experiment of function_IO
Reacting on IO
1685763545 3537075590 2609203750
Notifying experiment of function_IO
Reacting on IO
674413125 3612101396 4286511957
Notifying experiment of function_IO
Reacting on IO
674413125 3612101396 674410561
N



Min Length: (74, (19, 25))
Max Length: (80, (3, 1))
Mean Length: 76.69042316258351
Min Rate: 0
Max Rate: 2490368
Mean Rate: 25917.82222222222
Max Age: 743
Mean Age: 143.89
NOT: 117
NAND: 4
AND: 0
OR_N: 422
OR: 335
AND_N: 214
NOR: 0
XOR: 0
EQU: 0


Min Length: (73, (14, 27))
Max Length: (80, (1, 7))
Mean Length: 76.7119021134594
Min Rate: 0
Max Rate: 4980736
Mean Rate: 29154.272222222222
Max Age: 2263
Mean Age: 140.62
NOT: 132
NAND: 5
AND: 0
OR_N: 469
OR: 358
AND_N: 203
NOR: 0
XOR: 0
EQU: 0


Min Length: (73, (12, 22))
Max Length: (80, (0, 8))
Mean Length: 76.7741935483871
Min Rate: 0
Max Rate: 1212416
Mean Rate: 24483.847777777777
Max Age: 799
Mean Age: 140.8388888888889
NOT: 122
NAND: 9
AND: 0
OR_N: 408
OR: 313
AND_N: 183
NOR: 0
XOR: 0
EQU: 0


Min Length: (72, (6, 21))
Max Length: (82, (17, 2))
Mean Length: 76.80957683741649
Min Rate: 0
Max Rate: 4915200
Mean Rate: 27498.02111111111
Max Age: 747
Mean Age: 142.6288888888889
NOT: 124
NAND: 10
AND: 1
OR_N: 447
OR: 339
AND_N: 220
NOR: 

Notifying experiment of function_IO
Reacting on IO
587767476 3749978772 4287413940
Notifying experiment of function_IO
Reacting on IO
2536460189 2619593569 2670714877
Notifying experiment of function_IO
Reacting on IO
2536460189 2619593569 51121308
Notifying experiment of function_IO
Reacting on IO
3700319291 1163401711 594648004
Notifying experiment of function_IO
Reacting on IO
702959409 1728422279 1877471159
Notifying experiment of function_IO
Reacting on IO
702959409 1728422279 149048880
Notifying experiment of function_IO
Reacting on IO
79591629 1175645022 1186921439
Notifying experiment of function_IO
Reacting on IO
4080129900 4243475733 4294958973
Notifying experiment of function_IO
Reacting on IO
4080129900 4243475733 51483240
Notifying experiment of function_IO
Reacting on IO
3208014044 660440123 1086953251
Notifying experiment of function_IO
Reacting on IO
2414886713 1019351111 3220324223
Notifying experiment of function_IO
Reacting on IO
2414886713 1019351111 2200973112
Noti

1534592841 491172506 1602140123
Notifying experiment of function_IO
Reacting on IO
1534592841 491172506 1110967617
Notifying experiment of function_IO
Reacting on IO
4263672134 2768566020 31295161
Notifying experiment of function_IO
Reacting on IO
3338050551 2825385276 4009230335
Notifying experiment of function_IO
Reacting on IO
3338050551 2825385276 1183845059
Notifying experiment of function_IO
Reacting on IO
1497427396 2855330698 4218482638
Notifying experiment of function_IO
Reacting on IO
2453547334 3265207639 3535748439
Notifying experiment of function_IO
Reacting on IO
2453547334 3265207639 270540800
Notifying experiment of function_IO
Reacting on IO
145704526 3241058791 4149262769
Notifying experiment of function_IO
Reacting on IO
1011032809 1171802783 2111531775
Notifying experiment of function_IO
Reacting on IO
1011032809 1171802783 939728992
Notifying experiment of function_IO
Reacting on IO
2719850774 3969082314 4003455966
Notifying experiment of function_IO
Reacting on IO



Min Length: (73, (26, 29))
Max Length: (79, (0, 7))
Mean Length: 77.25055679287306
Min Rate: 0
Max Rate: 5046272
Mean Rate: 29868.784444444445
Max Age: 2131
Mean Age: 130.6988888888889
NOT: 118
NAND: 6
AND: 1
OR_N: 473
OR: 336
AND_N: 206
NOR: 0
XOR: 0
EQU: 0


Min Length: (73, (19, 25))
Max Length: (79, (0, 14))
Mean Length: 77.34149054505005
Min Rate: 0
Max Rate: 5111808
Mean Rate: 29555.943333333333
Max Age: 1158
Mean Age: 135.92888888888888
NOT: 118
NAND: 10
AND: 0
OR_N: 475
OR: 341
AND_N: 192
NOR: 0
XOR: 0
EQU: 0


Min Length: (73, (13, 22))
Max Length: (79, (0, 0))
Mean Length: 77.26807563959956
Min Rate: 0
Max Rate: 5111808
Mean Rate: 34298.78333333333
Max Age: 2035
Mean Age: 130.85222222222222
NOT: 113
NAND: 7
AND: 0
OR_N: 457
OR: 346
AND_N: 187
NOR: 0
XOR: 0
EQU: 0


Min Length: (73, (4, 14))
Max Length: (79, (6, 1))
Mean Length: 77.16592427616926
Min Rate: 0
Max Rate: 5111808
Mean Rate: 29138.45111111111
Max Age: 2487
Mean Age: 135.93
NOT: 148
NAND: 4
AND: 1
OR_N: 470
OR: 36

Notifying experiment of function_IO
Reacting on IO
4255915321 2067148568 4290658105
Notifying experiment of function_IO
Reacting on IO
1677506374 2124924851 2147481591
Notifying experiment of function_IO
Reacting on IO
1677506374 2124924851 22556740
Notifying experiment of function_IO
Reacting on IO
3159836695 384618584 1135130600
Notifying experiment of function_IO
Reacting on IO
1440051933 4074869631 4160091135
Notifying experiment of function_IO
Reacting on IO
1440051933 4074869631 85221504
Notifying experiment of function_IO
Reacting on IO
979870534 2847870210 3154095942
Notifying experiment of function_IO
Reacting on IO
3937039578 735485999 3959354623
Notifying experiment of function_IO
Reacting on IO
3937039578 735485999 3223868624
Notifying experiment of function_IO
Reacting on IO
1541735390 3785905946 2753231905
Notifying experiment of function_IO
Reacting on IO
3337682679 232410101 3489333239
Notifying experiment of function_IO
Reacting on IO
3337682679 232410101 3256923138
No

Notifying experiment of function_IO
Reacting on IO
3768867103 3372047695 3908918623
Notifying experiment of function_IO
Reacting on IO
3768867103 3372047695 536870928


Min Length: (73, (1, 2))
Max Length: (80, (1, 8))
Mean Length: 77.77060133630289
Min Rate: 0
Max Rate: 5111808
Mean Rate: 29373.22888888889
Max Age: 1966
Mean Age: 136.45888888888888
NOT: 117
NAND: 3
AND: 1
OR_N: 444
OR: 348
AND_N: 218
NOR: 1
XOR: 0
EQU: 0
Reacting on nor
Notifying experiment of function_IO
Reacting on IO
1145390904 1449226207 1449488383
Notifying experiment of function_IO
Reacting on IO
2460265714 2825608012 3136290814
Notifying experiment of function_IO
Reacting on IO
2460265714 2825608012 310682802
Notifying experiment of function_IO
Reacting on IO
4004401235 323507076 290566060
Notifying experiment of function_IO
Reacting on IO
669953622 3073548936 3086921438
Notifying experiment of function_IO
Reacting on IO
669953622 3073548936 13372502
Notifying experiment of function_IO
Reacting on IO
464126716 

Reacting on nor
Notifying experiment of function_IO
Reacting on IO
3604616504 1454473975 3606812671
Notifying experiment of function_IO
Reacting on IO
3111053775 3284293218 4226744303
Notifying experiment of function_IO
Reacting on IO
3111053775 3284293218 942451085
Notifying experiment of function_IO
Reacting on IO
2509670144 1678930019 1785297151
Notifying experiment of function_IO
Reacting on IO
2204279361 2649711302 2683298503
Notifying experiment of function_IO
Reacting on IO
2204279361 2649711302 33587201
Notifying experiment of function_IO
Reacting on IO
853915480 2358611437 3203776509
Notifying experiment of function_IO
Reacting on IO
913316565 104589133 914096093
Notifying experiment of function_IO
Reacting on IO
913316565 104589133 809506960
Notifying experiment of function_IO
Reacting on IO
3191975474 2074186788 1102991821
Notifying experiment of function_IO
Reacting on IO
886195978 263252842 1072885610
Notifying experiment of function_IO
Reacting on IO
886195978 263252842 8



Min Length: (73, (2, 7))
Max Length: (80, (3, 22))
Mean Length: 77.66518353726363
Min Rate: 0
Max Rate: 5111808
Mean Rate: 35949.09111111111
Max Age: 2223
Mean Age: 145.33
NOT: 135
NAND: 13
AND: 0
OR_N: 441
OR: 338
AND_N: 200
NOR: 0
XOR: 0
EQU: 0


Min Length: (73, (8, 0))
Max Length: (80, (1, 25))
Mean Length: 77.65551839464882
Min Rate: 0
Max Rate: 5177344
Mean Rate: 30464.066666666666
Max Age: 1253
Mean Age: 147.77666666666667
NOT: 124
NAND: 7
AND: 0
OR_N: 481
OR: 350
AND_N: 205
NOR: 0
XOR: 0
EQU: 0


Min Length: (74, (1, 2))
Max Length: (80, (9, 24))
Mean Length: 77.5033407572383
Min Rate: 0
Max Rate: 5111808
Mean Rate: 29399.993333333332
Max Age: 1719
Mean Age: 142.16555555555556
NOT: 118
NAND: 4
AND: 0
OR_N: 414
OR: 327
AND_N: 196
NOR: 0
XOR: 0
EQU: 0


Min Length: (72, (21, 17))
Max Length: (80, (9, 24))
Mean Length: 77.57988826815642
Min Rate: 0
Max Rate: 5111808
Mean Rate: 27227.446666666667
Max Age: 2195
Mean Age: 106.3411111111111
NOT: 63
NAND: 4
AND: 2
OR_N: 369
OR: 239
A



Min Length: (74, (10, 16))
Max Length: (80, (24, 24))
Mean Length: 77.36929922135707
Min Rate: 0
Max Rate: 5111808
Mean Rate: 31569.113333333335
Max Age: 1526
Mean Age: 144.63222222222223
NOT: 159
NAND: 7
AND: 1
OR_N: 508
OR: 392
AND_N: 250
NOR: 0
XOR: 0
EQU: 0


Min Length: (74, (1, 7))
Max Length: (80, (9, 24))
Mean Length: 77.42046718576196
Min Rate: 0
Max Rate: 5177344
Mean Rate: 28802.525555555556
Max Age: 2068
Mean Age: 143.9388888888889
NOT: 120
NAND: 6
AND: 1
OR_N: 426
OR: 314
AND_N: 198
NOR: 0
XOR: 0
EQU: 0


Min Length: (73, (3, 6))
Max Length: (80, (1, 18))
Mean Length: 77.40555555555555
Min Rate: 73
Max Rate: 5046272
Mean Rate: 34360.32
Max Age: 671
Mean Age: 151.31333333333333
NOT: 148
NAND: 6
AND: 2
OR_N: 491
OR: 361
AND_N: 241
NOR: 0
XOR: 0
EQU: 0


Min Length: (74, (2, 7))
Max Length: (80, (8, 23))
Mean Length: 77.23222222222222
Min Rate: 75
Max Rate: 2490368
Mean Rate: 25991.498888888887
Max Age: 1862
Mean Age: 160.88777777777779
NOT: 136
NAND: 4
AND: 1
OR_N: 457
OR:



Min Length: (73, (2, 14))
Max Length: (80, (0, 3))
Mean Length: 78.48826815642458
Min Rate: 0
Max Rate: 5177344
Mean Rate: 29899.825555555555
Max Age: 1212
Mean Age: 113.88888888888889
NOT: 87
NAND: 3
AND: 1
OR_N: 348
OR: 238
AND_N: 161
NOR: 0
XOR: 0
EQU: 0
Reacting on nor
Notifying experiment of function_IO
Reacting on IO
278675889 1011507073 1021206449
Notifying experiment of function_IO
Reacting on IO
4019415550 2321249269 4024138751
Notifying experiment of function_IO
Reacting on IO
4019415550 2321249269 1702889482
Notifying experiment of function_IO
Reacting on IO
3435414417 115469474 859552878
Notifying experiment of function_IO
Reacting on IO
3305704246 1545953901 3710741375
Notifying experiment of function_IO
Reacting on IO
3305704246 1545953901 2164787474
Notifying experiment of function_IO
Reacting on IO
4188413312 1448832284 4294933916
Notifying experiment of function_IO
Reacting on IO
3680600232 1320753528 3757677048
Notifying experiment of function_IO
Reacting on IO
3680

Notifying experiment of function_IO
Reacting on IO
1988203610 2886347741 4270571487
Notifying experiment of function_IO
Reacting on IO
1124427994 1474098473 1474166267
Notifying experiment of function_IO
Reacting on IO
1124427994 1474098473 67794
Notifying experiment of function_IO
Reacting on IO
100403829 1759347293 4194563466
Notifying experiment of function_IO
Reacting on IO
518802544 2294183531 2667478651
Notifying experiment of function_IO
Reacting on IO
518802544 2294183531 373295120
Notifying experiment of function_IO
Reacting on IO
771746171 2941735810 2952788987
Notifying experiment of function_IO
Reacting on IO
3018165362 3910856380 4227854078
Notifying experiment of function_IO
Reacting on IO
3018165362 3910856380 316997698
Notifying experiment of function_IO
Reacting on IO
302535306 2000110292 3992431989
Notifying experiment of function_IO
Reacting on IO
3552195014 174590286 3690609102
Notifying experiment of function_IO
Reacting on IO
3552195014 174590286 3516018816
Notify

Notifying experiment of function_IO
Reacting on IO
2781875496 2444658769 3052834169
Notifying experiment of function_IO
Reacting on IO
1951764011 2266827370 4150103659
Notifying experiment of function_IO
Reacting on IO
1951764011 2266827370 1883276289
Notifying experiment of function_IO
Reacting on IO
554944581 2258563305 3740022714
Notifying experiment of function_IO
Reacting on IO
291613717 1739359248 2012064789
Notifying experiment of function_IO
Reacting on IO
291613717 1739359248 272705541
Notifying experiment of function_IO
Reacting on IO
228207341 3879819428 4024138477
Notifying experiment of function_IO
Reacting on IO
2819835278 2563806570 3100890606
Notifying experiment of function_IO
Reacting on IO
2819835278 2563806570 537084036
Notifying experiment of function_IO
Reacting on IO
2629400426 1304113426 1665566869
Notifying experiment of function_IO
Reacting on IO
3512277103 781234844 4292459263
Notifying experiment of function_IO
Reacting on IO
3512277103 781234844 3511224419




Min Length: (73, (3, 5))
Max Length: (81, (19, 22))
Mean Length: 78.53957636566332
Min Rate: 0
Max Rate: 5177344
Mean Rate: 28610.535555555554
Max Age: 2353
Mean Age: 147.4
NOT: 130
NAND: 5
AND: 2
OR_N: 421
OR: 340
AND_N: 213
NOR: 0
XOR: 0
EQU: 0


Min Length: (74, (7, 4))
Max Length: (81, (10, 10))
Mean Length: 78.51839464882943
Min Rate: 0
Max Rate: 5177344
Mean Rate: 30633.18
Max Age: 1183
Mean Age: 141.64333333333335
NOT: 125
NAND: 5
AND: 3
OR_N: 463
OR: 333
AND_N: 206
NOR: 0
XOR: 0
EQU: 0
Reacting on nor
Notifying experiment of function_IO
Reacting on IO
2356473479 3691674329 3699310303
Notifying experiment of function_IO
Reacting on IO
221864141 1512778344 1597729517
Notifying experiment of function_IO
Reacting on IO
221864141 1512778344 84951173
Notifying experiment of function_IO
Reacting on IO
1733394006 612342541 2561573289
Notifying experiment of function_IO
Reacting on IO
2869036688 3149095271 3149096951
Notifying experiment of function_IO
Reacting on IO
2869036688 314909

Notifying experiment of function_IO
Reacting on IO
1875406069 475461509 2145320949
Notifying experiment of function_IO
Reacting on IO
1813315822 3430869882 3967806462
Notifying experiment of function_IO
Reacting on IO
1813315822 3430869882 536936580
Notifying experiment of function_IO
Reacting on IO
4067986145 1925129729 226981150
Notifying experiment of function_IO
Reacting on IO
1298926755 893549893 2104393191
Notifying experiment of function_IO
Reacting on IO
1298926755 893549893 1210843298
Notifying experiment of function_IO
Reacting on IO
4120172113 570107259 4127192955
Notifying experiment of function_IO
Reacting on IO
2256133848 1939905673 4160347865
Notifying experiment of function_IO
Reacting on IO
2256133848 1939905673 2220442192
Notifying experiment of function_IO
Reacting on IO
1304406638 3383338502 2990560657
Notifying experiment of function_IO
Reacting on IO
2755050792 1977519497 4127112617
Notifying experiment of function_IO
Reacting on IO
2755050792 1977519497 214959312

Notifying experiment of function_IO
Reacting on IO
1825970674 3013809142 4294384630
Notifying experiment of function_IO
Reacting on IO
3044563771 1315025726 4286184255
Notifying experiment of function_IO
Reacting on IO
3044563771 1315025726 2971158529
Notifying experiment of function_IO
Reacting on IO
1194226238 3555263147 3100741057
Notifying experiment of function_IO
Reacting on IO
422062268 22198099 427474943
Notifying experiment of function_IO
Reacting on IO
422062268 22198099 405276844
Reacting on nor
Notifying experiment of function_IO
Reacting on IO
4257950030 370198430 4292603870
Notifying experiment of function_IO
Reacting on IO
83131028 3030367127 3036674967
Notifying experiment of function_IO
Reacting on IO
83131028 3030367127 6307840
Notifying experiment of function_IO
Reacting on IO
2746543890 322411337 1548423405
Notifying experiment of function_IO
Reacting on IO
820426162 1254101001 2061940155
Notifying experiment of function_IO
Reacting on IO
820426162 1254101001 807839



Min Length: (75, (4, 0))
Max Length: (81, (6, 19))
Mean Length: 79.15590200445435
Min Rate: 0
Max Rate: 5177344
Mean Rate: 29481.518888888888
Max Age: 750
Mean Age: 130.69555555555556
NOT: 133
NAND: 4
AND: 1
OR_N: 405
OR: 322
AND_N: 214
NOR: 1
XOR: 0
EQU: 0


Min Length: (75, (0, 3))
Max Length: (81, (4, 8))
Mean Length: 79.14905450500557
Min Rate: 0
Max Rate: 5177344
Mean Rate: 29716.044444444444
Max Age: 721
Mean Age: 146.99444444444444
NOT: 134
NAND: 6
AND: 0
OR_N: 468
OR: 360
AND_N: 230
NOR: 0
XOR: 0
EQU: 0
Reacting on nor
Reacting on nor
Reacting on nor
Notifying experiment of function_IO
Reacting on IO
1145838258 1373262730 1440635834
Notifying experiment of function_IO
Reacting on IO
2844609589 389971656 3216998141
Notifying experiment of function_IO
Reacting on IO
2844609589 389971656 2827026485
Notifying experiment of function_IO
Reacting on IO
527993861 3451175349 3757886901
Notifying experiment of function_IO
Reacting on IO
882368371 2154133709 3036151807
Notifying experim



Min Length: (75, (3, 16))
Max Length: (81, (0, 24))
Mean Length: 78.72048997772829
Min Rate: 0
Max Rate: 5242880
Mean Rate: 28591.25
Max Age: 1868
Mean Age: 137.67444444444445
NOT: 93
NAND: 7
AND: 0
OR_N: 375
OR: 288
AND_N: 192
NOR: 0
XOR: 0
EQU: 0


Min Length: (75, (2, 18))
Max Length: (81, (5, 9))
Mean Length: 78.70444444444445
Min Rate: 75
Max Rate: 5111808
Mean Rate: 28389.452222222222
Max Age: 685
Mean Age: 144.7188888888889
NOT: 121
NAND: 4
AND: 0
OR_N: 413
OR: 324
AND_N: 215
NOR: 0
XOR: 0
EQU: 0


Min Length: (76, (0, 22))
Max Length: (82, (25, 17))
Mean Length: 79.08053691275168
Min Rate: 0
Max Rate: 5242880
Mean Rate: 24804.65111111111
Max Age: 2512
Mean Age: 108.61888888888889
NOT: 75
NAND: 3
AND: 0
OR_N: 326
OR: 221
AND_N: 137
NOR: 0
XOR: 0
EQU: 0
Reacting on nor
Notifying experiment of function_IO
Reacting on IO
2599019648 2126806243 4277009635
Notifying experiment of function_IO
Reacting on IO
3284479613 1961145980 4159028861
Notifying experiment of function_IO
Reacting

Notifying experiment of function_IO
Reacting on IO
4185941171 2556329799 4192108535
Notifying experiment of function_IO
Reacting on IO
283839790 961293128 971976558
Notifying experiment of function_IO
Reacting on IO
283839790 961293128 10683430
Notifying experiment of function_IO
Reacting on IO
3427819893 496044729 867147402
Notifying experiment of function_IO
Reacting on IO
1977989824 2494155376 4126014192
Notifying experiment of function_IO
Reacting on IO
1977989824 2494155376 1631858816
Reacting on nor
Notifying experiment of function_IO
Reacting on IO
3173454240 3396065505 4285520353
Notifying experiment of function_IO
Reacting on IO
2538418424 3166088201 3221208313
Notifying experiment of function_IO
Reacting on IO
2538418424 3166088201 55120112
Notifying experiment of function_IO
Reacting on IO
1733130125 3598608613 2561837170
Notifying experiment of function_IO
Reacting on IO
1170221046 1274325690 1341435902
Notifying experiment of function_IO
Reacting on IO
1170221046 127432569

Notifying experiment of function_IO
Reacting on IO
2518197339 30375861 2548039679
Notifying experiment of function_IO
Reacting on IO
2518197339 30375861 2517663818
Notifying experiment of function_IO
Reacting on IO
4255289334 474712239 4260073471
Notifying experiment of function_IO
Reacting on IO
683968461 22595513 702334973
Notifying experiment of function_IO
Reacting on IO
683968461 22595513 679739460
Notifying experiment of function_IO
Reacting on IO
3885477937 4010493627 409489358
Notifying experiment of function_IO
Reacting on IO
166271473 2775461213 2918137341
Notifying experiment of function_IO
Reacting on IO
166271473 2775461213 142676128
Notifying experiment of function_IO
Reacting on IO
734155258 2031472473 2077745147
Notifying experiment of function_IO
Reacting on IO
2290614240 3472898121 3481796585
Notifying experiment of function_IO
Reacting on IO
2290614240 3472898121 8898464
Notifying experiment of function_IO
Reacting on IO
1539099853 2953628362 2755867442
Notifying exp

Reacting on nor
Notifying experiment of function_IO
Reacting on IO
3127865733 2757141107 3196056567
Notifying experiment of function_IO
Reacting on IO
2979192523 2611957065 3149914059
Notifying experiment of function_IO
Reacting on IO
2979192523 2611957065 537956994
Notifying experiment of function_IO
Reacting on IO
435196185 1661954355 3859771110
Notifying experiment of function_IO
Reacting on IO
2806880920 782232293 2950691581
Notifying experiment of function_IO
Reacting on IO
2806880920 782232293 2168459288
Notifying experiment of function_IO
Reacting on IO
985526259 2397288258 3204448243
Notifying experiment of function_IO
Reacting on IO
2286386287 2660544888 2664955263
Notifying experiment of function_IO
Reacting on IO
2286386287 2660544888 4410375
Notifying experiment of function_IO
Reacting on IO
3752234705 1121860332 542732590
Notifying experiment of function_IO
Reacting on IO
2890092861 2159518192 2901930493
Notifying experiment of function_IO
Reacting on IO
2890092861 2159518

Notifying experiment of function_IO
Reacting on IO
3143051625 347483192 2873164097
Notifying experiment of function_IO
Reacting on IO
2229098839 2407992677 2065868456
Notifying experiment of function_IO
Reacting on IO
697629997 1195690818 1876229999
Notifying experiment of function_IO
Reacting on IO
697629997 1195690818 680539181
Notifying experiment of function_IO
Reacting on IO
1165403563 1363897604 1434447279
Notifying experiment of function_IO
Reacting on IO
2733662218 839984136 3002164234
Notifying experiment of function_IO
Reacting on IO
2733662218 839984136 2162180098
Notifying experiment of function_IO
Reacting on IO
1916095643 1691523103 2378871652
Notifying experiment of function_IO
Reacting on IO
525935216 997488410 1065187194
Notifying experiment of function_IO
Reacting on IO
525935216 997488410 67698784


Min Length: (76, (13, 29))
Max Length: (83, (16, 18))
Mean Length: 79.99109131403118
Min Rate: 0
Max Rate: 5242880
Mean Rate: 29094.864444444444
Max Age: 2119
Mean Age: 1

Notifying experiment of function_IO
Reacting on IO
2366832717 1995889863 4294433999
Notifying experiment of function_IO
Reacting on IO
1193575388 3936646244 4020566012
Notifying experiment of function_IO
Reacting on IO
1373213653 3135326970 4227455999
Notifying experiment of function_IO
Reacting on IO
4092461453 409548279 4226758143
Notifying experiment of function_IO
Reacting on IO
2070361065 2805924114 4286526459
Notifying experiment of function_IO
Reacting on IO
1926266588 2949050801 4292280317
Notifying experiment of function_IO
Reacting on IO
3294770000 3358600511 3430039423
Notifying experiment of function_IO
Reacting on IO
2313127923 2065403431 4225732599
Notifying experiment of function_IO
Reacting on IO
678588078 535330418 1073381118
Notifying experiment of function_IO
Reacting on IO
745099849 1678353639 1818883823


Min Length: (77, (14, 15))
Max Length: (82, (6, 24))
Mean Length: 79.80422691879866
Min Rate: 0
Max Rate: 5308416
Mean Rate: 29393.324444444443
Max Age: 770
Mean 

Notifying experiment of function_IO
Reacting on IO
1112057587 1434774773 1473048311
Notifying experiment of function_IO
Reacting on IO
2920733909 589036717 2938109181
Notifying experiment of function_IO
Reacting on IO
2920733909 589036717 2349072464
Notifying experiment of function_IO
Reacting on IO
3200807908 1231424061 1094159387
Notifying experiment of function_IO
Reacting on IO
3178937090 4004034010 4294623194
Notifying experiment of function_IO
Reacting on IO
3178937090 4004034010 290589184
Notifying experiment of function_IO
Reacting on IO
2769934501 2323993106 2946357943
Notifying experiment of function_IO
Reacting on IO
1033010384 228482038 1033797622
Notifying experiment of function_IO
Reacting on IO
1033010384 228482038 805315584
Notifying experiment of function_IO
Reacting on IO
3475801788 2840720354 819165507
Notifying experiment of function_IO
Reacting on IO
110744737 3575789238 3619419831
Notifying experiment of function_IO
Reacting on IO
110744737 3575789238 43630593
Not



Min Length: (76, (7, 27))
Max Length: (82, (18, 11))
Mean Length: 79.75473801560759
Min Rate: 0
Max Rate: 5308416
Mean Rate: 30168.687777777777
Max Age: 1260
Mean Age: 143.54111111111112
NOT: 139
NAND: 6
AND: 0
OR_N: 468
OR: 371
AND_N: 216
NOR: 0
XOR: 0
EQU: 0


Min Length: (76, (16, 20))
Max Length: (88, (25, 29))
Mean Length: 79.70489977728285
Min Rate: 0
Max Rate: 5242880
Mean Rate: 28137.352222222224
Max Age: 659
Mean Age: 132.22333333333333
NOT: 105
NAND: 4
AND: 0
OR_N: 411
OR: 333
AND_N: 203
NOR: 0
XOR: 0
EQU: 0


Min Length: (77, (5, 15))
Max Length: (82, (27, 0))
Mean Length: 79.61
Min Rate: 77
Max Rate: 665600
Mean Rate: 22195.795555555556
Max Age: 1041
Mean Age: 111.94666666666667
NOT: 97
NAND: 5
AND: 0
OR_N: 295
OR: 224
AND_N: 134
NOR: 0
XOR: 0
EQU: 0


Min Length: (76, (9, 7))
Max Length: (83, (28, 6))
Mean Length: 79.57730812013348
Min Rate: 0
Max Rate: 2588672
Mean Rate: 26318.237777777777
Max Age: 2265
Mean Age: 147.59555555555556
NOT: 128
NAND: 12
AND: 1
OR_N: 425
OR:

Notifying experiment of function_IO
Reacting on IO
2097004288 798840347 2147352347
Notifying experiment of function_IO
Reacting on IO
1611142254 662999984 1737267198
Notifying experiment of function_IO
Reacting on IO
1611142254 662999984 1074267214
Notifying experiment of function_IO
Reacting on IO
4107192724 105704968 187774571
Notifying experiment of function_IO
Reacting on IO
2643528731 411927234 2644348635
Notifying experiment of function_IO
Reacting on IO
2643528731 411927234 2232421401
Notifying experiment of function_IO
Reacting on IO
4146060214 1978958055 4160740343
Notifying experiment of function_IO
Reacting on IO
1815808313 3852471350 3988458815
Notifying experiment of function_IO
Reacting on IO
1815808313 3852471350 135987465
Notifying experiment of function_IO
Reacting on IO
4192957611 450822610 102009684
Notifying experiment of function_IO
Reacting on IO
336349271 2259816827 2529057151
Notifying experiment of function_IO
Reacting on IO
336349271 2259816827 269240324
Notif

Reacting on nor
Notifying experiment of function_IO
Reacting on IO
1299415087 4288853852 4294162303
Notifying experiment of function_IO
Reacting on IO
550392671 4045833676 4058961887
Notifying experiment of function_IO
Reacting on IO
550392671 4045833676 13128211
Notifying experiment of function_IO
Reacting on IO
3388998801 1970885642 905968494
Notifying experiment of function_IO
Reacting on IO
1547810542 3556887150 3695312622
Notifying experiment of function_IO
Reacting on IO
1547810542 3556887150 138425472
Notifying experiment of function_IO
Reacting on IO
1391144829 333695290 1407971199
Notifying experiment of function_IO
Reacting on IO
3722630344 1624008763 4260297979
Notifying experiment of function_IO
Reacting on IO
3722630344 1624008763 2636289216
Notifying experiment of function_IO
Reacting on IO
3366973906 1194994493 927993389
Notifying experiment of function_IO
Reacting on IO
2687806050 794148107 2943729515
Notifying experiment of function_IO
Reacting on IO
2687806050 7941481

Reacting on nor
Reacting on nor


Min Length: (75, (1, 15))
Max Length: (82, (7, 1))
Mean Length: 79.87527839643653
Min Rate: 0
Max Rate: 41984
Mean Rate: 24098.14888888889
Max Age: 797
Mean Age: 161.78666666666666
NOT: 133
NAND: 6
AND: 1
OR_N: 492
OR: 352
AND_N: 221
NOR: 0
XOR: 0
EQU: 0
Reacting on nor


Min Length: (75, (1, 12))
Max Length: (82, (1, 3))
Mean Length: 79.80957683741649
Min Rate: 0
Max Rate: 5177344
Mean Rate: 30620.272222222222
Max Age: 2032
Mean Age: 150.24777777777777
NOT: 138
NAND: 6
AND: 2
OR_N: 443
OR: 347
AND_N: 222
NOR: 0
XOR: 0
EQU: 0


Min Length: (76, (1, 6))
Max Length: (83, (28, 16))
Mean Length: 79.88182831661092
Min Rate: 0
Max Rate: 5177344
Mean Rate: 29498.446666666667
Max Age: 812
Mean Age: 135.07222222222222
NOT: 115
NAND: 2
AND: 1
OR_N: 422
OR: 321
AND_N: 216
NOR: 0
XOR: 0
EQU: 0


Min Length: (75, (4, 13))
Max Length: (91, (12, 16))
Mean Length: 79.92865105908584
Min Rate: 0
Max Rate: 5242880
Mean Rate: 30766.428888888888
Max Age: 1348
Mean Age: 146



Min Length: (76, (13, 23))
Max Length: (83, (11, 28))
Mean Length: 80.09608938547485
Min Rate: 0
Max Rate: 5373952
Mean Rate: 35798.926666666666
Max Age: 879
Mean Age: 128.75333333333333
NOT: 106
NAND: 10
AND: 2
OR_N: 436
OR: 312
AND_N: 198
NOR: 0
XOR: 0
EQU: 0


Min Length: (76, (1, 1))
Max Length: (83, (16, 13))
Mean Length: 80.09121245828699
Min Rate: 0
Max Rate: 5177344
Mean Rate: 30212.48222222222
Max Age: 1710
Mean Age: 147.3411111111111
NOT: 119
NAND: 8
AND: 1
OR_N: 429
OR: 331
AND_N: 205
NOR: 0
XOR: 0
EQU: 0


Min Length: (76, (5, 1))
Max Length: (83, (13, 14))
Mean Length: 80.2769744160178
Min Rate: 0
Max Rate: 5242880
Mean Rate: 29061.431111111113
Max Age: 1137
Mean Age: 147.42444444444445
NOT: 84
NAND: 10
AND: 1
OR_N: 407
OR: 329
AND_N: 184
NOR: 0
XOR: 0
EQU: 0
Reacting on nor
Notifying experiment of function_IO
Reacting on IO
698082720 3676830889 4223663529
Notifying experiment of function_IO
Reacting on IO
226066258 1971801211 2113896315
Notifying experiment of function_

In [17]:
experiment.display_statistics()

Min Length: (48, (10, 6))
Max Length: (59, (14, 18))
Mean Length: 54.33
Min Rate: 51
Max Rate: 3520
Mean Rate: 447.8544444444444
Max Age: 1339
Mean Age: 111.11
NOT: 154
NAND: 2
AND: 0
OR_N: 47
OR: 494
AND_N: 0
NOR: 1
XOR: 0
EQU: 0


In [20]:
"""
A script to count the mutation types
"""

c = 0
ins = 0
d = 0

for i in range(len(em.mutations)):
    
    if em.mutations[i][0] == "C":
        c += 1
    elif em.mutations[i][0] == "I":
        ins += 1
    elif em.mutations[i][0] == "D":
        d += 1
        
print("Copy: " + str(c))
print("Insert: " + str(ins))
print("Delete: " + str(d))
print("TOTAL: " + str(c+d+ins))

Copy: 43
Insert: 14
Delete: 10
TOTAL: 67


In [21]:
em.mutations

[['I', 47, 19, 6],
 ['C', 27, 15, 11],
 ['C', 2, 0, 15],
 ['C', 12, 19, 16],
 ['I', 9, 27, 17],
 ['C', 39, 19, 23],
 ['I', 22, 19, 23],
 ['D', 37, 24],
 ['C', 19, 2, 26],
 ['C', 28, 26, 30],
 ['I', 6, 19, 32],
 ['I', 8, 17, 34],
 ['I', 45, 19, 38],
 ['C', 26, 0, 44],
 ['C', 33, 5, 48],
 ['I', 38, 13, 49],
 ['I', 1, 26, 51],
 ['C', 7, 5, 63],
 ['I', 32, 14, 64],
 ['C', 24, 0, 67],
 ['D', 34, 67],
 ['C', 4, 23, 68],
 ['C', 19, 23, 69],
 ['C', 11, 3, 74],
 ['C', 44, 16, 80],
 ['C', 44, 12, 81],
 ['C', 20, 27, 83],
 ['C', 26, 19, 84],
 ['C', 19, 18, 87],
 ['C', 51, 7, 93],
 ['D', 30, 95],
 ['C', 3, 18, 97],
 ['C', 16, 22, 100],
 ['D', 13, 104],
 ['C', 31, 27, 109],
 ['C', 38, 18, 113],
 ['I', 14, 21, 117],
 ['C', 42, 4, 123],
 ['D', 46, 124],
 ['C', 28, 18, 127],
 ['D', 19, 128],
 ['C', 20, 14, 132],
 ['C', 16, 6, 137],
 ['C', 26, 15, 142],
 ['C', 43, 14, 145],
 ['C', 14, 8, 149],
 ['C', 23, 16, 152],
 ['D', 44, 158],
 ['C', 7, 25, 162],
 ['C', 45, 9, 164],
 ['D', 9, 179],
 ['C', 41, 7, 18

In [19]:
"""
A script to test the evolved organism
"""

world = World(1,notify_ = True)
world.place_custom(em.original_memory)
world.schedule()


EMULATOR AT POSITION (0, 0) COMPUTED NOR
INPUT: [3498750375  280045140]
OUTPUT:  793004040


KeyboardInterrupt: 

In [ ]:
"""
A Script to recreate the program from the ancestor and the mutations
"""

start = em.ancestor.copy()
mutations = em.mutations.copy()

for mutation in mutations:
    
    if mutation[0] == "C":
        start[mutation[1]] = mutation[2]
        
    if mutation[0] == "I":
        start.insert(mutation[1], mutation[2])
        
    if mutation[0] == "D":
        del start[mutation[1]]
        
print(start == experiment.first_specimen)

In [12]:
"""
A list of all the mutations which led to the final organism
"""
mutations

[['C', 2, 23, 0],
 ['C', 44, 19, 0],
 ['C', 44, 9, 19],
 ['C', 60, 1, 19],
 ['C', 50, 3, 45],
 ['I', 63, 19, 61],
 ['C', 60, 4, 81],
 ['C', 55, 24, 82],
 ['C', 58, 25, 98],
 ['C', 20, 0, 121],
 ['I', 64, 19, 136],
 ['C', 2, 26, 143],
 ['I', 8, 19, 165],
 ['I', 51, 3, 169],
 ['C', 61, 1, 173],
 ['C', 58, 18, 176],
 ['C', 59, 20, 179],
 ['C', 36, 2, 186],
 ['D', 29, 198],
 ['I', 64, 19, 202],
 ['C', 8, 8, 214],
 ['C', 61, 12, 214],
 ['C', 60, 20, 221],
 ['I', 43, 16, 223],
 ['C', 0, 10, 226],
 ['C', 60, 15, 232],
 ['C', 62, 5, 235],
 ['I', 63, 18, 259],
 ['C', 3, 5, 263],
 ['I', 20, 6, 275],
 ['C', 46, 17, 308],
 ['C', 64, 16, 309],
 ['C', 3, 14, 311],
 ['C', 62, 19, 320],
 ['I', 29, 4, 323],
 ['D', 24, 325],
 ['C', 56, 5, 335],
 ['I', 26, 17, 335],
 ['C', 17, 13, 340],
 ['I', 47, 15, 340]]

XOR -> EQU:

340 Generations

Mutations:
Copy: 27
Insert: 11
Delete: 2
TOTAL: 40

EQU of length 79

Does contain the "move" instruction

[10, 20, 26, 14, 21, 2, 18, 0, 8, 18, 2, 7, 0, 7, 2, 8, 14, 13, 12, 2, 6, 7, 0, 6, 6, 0, 17, 15, 2, 4, 6, 7, 2, 8, 6, 0, 6, 2, 14, 2, 12, 2, 7, 0, 6, 16, 1, 15, 17, 0, 15, 2, 8, 6, 3, 3, 15, 0, 5, 0, 24, 18, 20, 15, 19, 5, 16, 20, 19, 19, 19, 19, 25, 2, 0, 17, 21, 0, 1]

In [13]:
"""
Recreating the program at all generations

After running this script, lineage[i] contains the program at generation i,
generation 0 being the original ancestor, generation 1 its child and so on
"""

generations = mutations[-1][-1] + 1

start = em.ancestor.copy()
mutations = em.mutations.copy()

temp = start.copy()

lineage = [start]

for i in range(generations):
    
    for mutation in mutations:
        
        if mutation[-1] == i:
            
            if mutation[0] == "C":
                temp[mutation[1]] = mutation[2]
        
            if mutation[0] == "I":
                temp.insert(mutation[1], mutation[2])
        
            if mutation[0] == "D":
                del temp[mutation[1]]
                
    lineage.append(temp.copy())

In [14]:
for specimen in lineage:
    print(specimen)
    print("\n")

[16, 20, 2, 0, 21, 2, 18, 0, 18, 2, 7, 0, 7, 2, 8, 14, 2, 12, 2, 7, 0, 6, 1, 6, 0, 15, 2, 6, 1, 7, 2, 8, 6, 0, 6, 2, 14, 2, 12, 2, 7, 0, 6, 1, 6, 0, 15, 2, 8, 6, 1, 15, 0, 18, 0, 6, 0, 6, 1, 6, 2, 20, 19, 25, 2, 0, 17, 21, 0, 1]


[16, 20, 23, 0, 21, 2, 18, 0, 18, 2, 7, 0, 7, 2, 8, 14, 2, 12, 2, 7, 0, 6, 1, 6, 0, 15, 2, 6, 1, 7, 2, 8, 6, 0, 6, 2, 14, 2, 12, 2, 7, 0, 6, 1, 19, 0, 15, 2, 8, 6, 1, 15, 0, 18, 0, 6, 0, 6, 1, 6, 2, 20, 19, 25, 2, 0, 17, 21, 0, 1]


[16, 20, 23, 0, 21, 2, 18, 0, 18, 2, 7, 0, 7, 2, 8, 14, 2, 12, 2, 7, 0, 6, 1, 6, 0, 15, 2, 6, 1, 7, 2, 8, 6, 0, 6, 2, 14, 2, 12, 2, 7, 0, 6, 1, 19, 0, 15, 2, 8, 6, 1, 15, 0, 18, 0, 6, 0, 6, 1, 6, 2, 20, 19, 25, 2, 0, 17, 21, 0, 1]


[16, 20, 23, 0, 21, 2, 18, 0, 18, 2, 7, 0, 7, 2, 8, 14, 2, 12, 2, 7, 0, 6, 1, 6, 0, 15, 2, 6, 1, 7, 2, 8, 6, 0, 6, 2, 14, 2, 12, 2, 7, 0, 6, 1, 19, 0, 15, 2, 8, 6, 1, 15, 0, 18, 0, 6, 0, 6, 1, 6, 2, 20, 19, 25, 2, 0, 17, 21, 0, 1]


[16, 20, 23, 0, 21, 2, 18, 0, 18, 2, 7, 0, 7, 2, 8, 14, 2, 12, 2, 7, 0

In [4]:
world = World(2, notify_ = False, cm_prob = 0, ins_prob = 0, del_prob = 0)
world.place_custom([10, 20, 26, 14, 21, 2, 18, 0, 8, 18, 2, 7, 0, 7, 2, 8, 14, 13, 12, 2, 6, 7, 0, 6, 6, 0, 17, 15, 2, 4, 6, 7, 2, 8, 6, 0, 6, 2, 14, 2, 12, 2, 7, 0, 6, 16, 1, 15, 17, 0, 15, 2, 8, 6, 3, 3, 15, 0, 5, 0, 24, 18, 20, 15, 19, 5, 16, 20, 19, 19, 19, 19, 25, 2, 0, 17, 21, 0, 1])

In [6]:
world.rates

array([[   79, 20224],
       [  316,   316]])

In [67]:
experiment = Experiment(start_organism="not", target_function="nor", ins_prob = 0,\
                        del_prob = 0, cm_prob = 0,N=3, stat_cycles = 100, instruction_set = "custom")

In [68]:
experiment.world.pool.get()

array([[<DigitalOrganism.CPUEmulator object at 0x000002AE875879A0>,
      dtype=object)

In [69]:
experiment.world.rates

array([[50, 50, 50],
       [50, 50, 50],
       [50, 50, 50]])

In [226]:
killer = [16, 20, 2, 0, 21] + [2] + [18, 0, 18, 2, 7, 0, 7, 2, 8, 14, 2, 12, 2, 7, 0, 6, 1, 6, 0, 15,2, 6, 1, 7, 2, 8, 6, 0, 6, 2, 14, 2, 12, 2, 7, 0, 6, 1, 6, 0, 15, 2, 8, 6, 1, 15, 0, 18, 0]\
                            + [6, 0, 6, 1, 6, 2] + [20, 19, 25, 2, 0, 17, 21, 0, 1]

killer = [10, 20, 26, 14, 21, 2, 18, 0, 8, 18, 2, 7, 0, 7, 2, 8, 14, 13, 12, 2, 6, 7, 0, 6, 6, 0, 17, 15, 2, 4, 6, 7, 2, 8, 6, 0, 6, 2, 14, 2, 12, 2, 7, 0, 6, 16, 1, 15, 17, 0, 15, 2, 8, 6, 3, 3, 15, 0, 5, 0, 24, 18, 20, 15, 19, 5, 16, 20, 19, 19, 19, 19, 25, 2, 0, 17, 21, 0, 1]

In [10]:
world = World(2, notify_ = False, cm_prob = 0, ins_prob = 0, del_prob = 0)
world.place_custom([0,1,1,2])